diff --git a/drivers/media/platform/apple/avd/avd-h264.c b/drivers/media/platform/apple/avd/avd-h264.c
index 7309151aab1d91..3e0da7fd2a6aed 100644
--- a/drivers/media/platform/apple/avd/avd-h264.c
+++ b/drivers/media/platform/apple/avd/avd-h264.c
@@ -100,14 +100,14 @@ static void stream_refs(struct avd_ctx *ctx, struct avd_h264_run *run)
 	struct avd_h264_ctx *h264_ctx = ctx->priv;
 	struct avd_dev *avd = ctx->dev;
 
-	push(INST_DMA2, "cm3_dma_config_6");
+	push(0, "");
 	pusha(h264_ctx->bufs.pps_tile[4].addr, "hdr_9c_pps_tile_addr_lsb8", 7);
 	pusha(run->addresses.sps, "hdr_bc_sps_tile_addr_lsb8", 0);
 
-	push(INST_DMA3, "cm3_dma_config_7");
-	push(INST_DMA3, "cm3_dma_config_8");
-	push(INST_DMA3, "cm3_dma_config_9");
-	push(INST_DMA3, "cm3_dma_config_a");
+	push(0, "");
+	push(0, "");
+	push(0, "");
+	push(0, "");
 
 	for (int i = 0; i < ARRAY_SIZE(decode->dpb); i++) {
 		if (!(dpb[i].flags & V4L2_H264_DPB_ENTRY_FLAG_VALID))
@@ -261,16 +261,16 @@ static void stream_hdr(struct avd_ctx *ctx, struct avd_h264_run *run)
 			     !(avd->variant->quirks & AVD_QUIRK_NO_PIPE_STATE)),
 	     "hdr_58_const_3a");
 
-	push(INST_DMA2, "cm3_dma_config_1");
-	push(INST_DMA1, "cm3_dma_config_2");
+	push(0, "");
+	push(0, "");
 
 	if (avd->variant->revision == 3)
 		push(0, "zero");
 
 	pusha(h264_ctx->bufs.pps_tile[0].addr, "hdr_9c_pps_tile_addr_lsb8", 0);
 
-	push(INST_DMA2, "cm3_dma_config_3");
-	push(INST_DMA2, "cm3_dma_config_4");
+	push(0, "");
+	push(0, "");
 
 	if (avd->variant->revision == 3)
 		push(0, "zero");
@@ -280,7 +280,7 @@ static void stream_hdr(struct avd_ctx *ctx, struct avd_h264_run *run)
 	pusha(h264_ctx->bufs.pps_tile[1].addr, "hdr_9c_pps_tile_addr_lsb8", 1);
 	pusha(h264_ctx->bufs.pps_tile[2].addr, "hdr_9c_pps_tile_addr_lsb8", 2);
 	pusha(h264_ctx->bufs.pps_tile[3].addr, "hdr_9c_pps_tile_addr_lsb8", 3);
-	push(INST_DMA3, "cm3_dma_config_5");
+	push(0, "");
 
 	push_rvra(avd, ctx, run->addresses.rvra, ctx->rvra.offsets);
 
diff --git a/drivers/media/platform/apple/avd/avd-hevc.c b/drivers/media/platform/apple/avd/avd-hevc.c
index 73aad4513cbd80..bf5bc366640c96 100644
--- a/drivers/media/platform/apple/avd/avd-hevc.c
+++ b/drivers/media/platform/apple/avd/avd-hevc.c
@@ -102,15 +102,15 @@ static void stream_refs(struct avd_ctx *ctx, struct avd_hevc_run *run)
 
 	dst = vb2_to_avd_decoded_buf(&run->base.bufs.dst->vb2_buf);
 
-	push(INST_DMA2, "cm3_dma_config_6");
+	push(0, "");
 	pusha(hevc_ctx->bufs.pps_tile[1].addr, "hdr_9c_pps_tile_addr_lsb8", 7);
 	pusha(run->addresses.sps, "hdr_bc_sps_tile_addr_lsb8",
 	      sl->slice_pic_order_cnt);
 
-	push(INST_DMA3, "cm3_dma_config_7");
-	push(INST_DMA3, "cm3_dma_config_8");
-	push(INST_DMA3, "cm3_dma_config_9");
-	push(INST_DMA3, "cm3_dma_config_a");
+	push(0, "");
+	push(0, "");
+	push(0, "");
+	push(0, "");
 
 	for (int i = 0; i < decode->num_active_dpb_entries; i++) {
 		dpb = &decode->dpb[i];
@@ -347,14 +347,14 @@ static void set_header(struct avd_ctx *ctx, struct avd_hevc_run *run)
 			     !(avd->variant->quirks & AVD_QUIRK_NO_PIPE_STATE)),
 	     "hdr_98_const_30");
 
-	push(INST_DMA2, "cm3_dma_config_1");
-	push(INST_DMA1, "cm3_dma_config_2");
+	push(0, "");
+	push(0, "");
 
 	if (avd->variant->revision == 3)
 		push(0, "zero");
 
-	push(INST_DMA1, "cm3_dma_config_2");
-	push(INST_DMA2, "cm3_dma_config_3");
+	push(0, "");
+	push(0, "");
 
 	if (avd->variant->revision == 3)
 		push(0, "zero");
@@ -382,7 +382,7 @@ static void set_header(struct avd_ctx *ctx, struct avd_hevc_run *run)
 		pusha(0, "", 9);
 	}
 
-	push(INST_DMA3, "cm3_dma_config_5");
+	push(0, "");
 
 	push_rvra(avd, ctx, run->addresses.rvra, ctx->rvra.offsets);
 
diff --git a/drivers/media/platform/apple/avd/avd-inst.h b/drivers/media/platform/apple/avd/avd-inst.h
index 8ee19f65e768e5..41815339b8a06f 100644
--- a/drivers/media/platform/apple/avd/avd-inst.h
+++ b/drivers/media/platform/apple/avd/avd-inst.h
@@ -7,11 +7,6 @@
 
 #include "avd.h"
 
-/* i have no clue what this is */
-#define INST_DMA1 0 /* (0x14 << 16 | 0x14) */
-#define INST_DMA2 0 /* (0x4000000 | INST_DMA1) */
-#define INST_DMA3 0 /* (0x07 << 16 | 0x07) */
-
 #define AVD_OP_EXEC			FIELD_PREP(GENMASK(31, 24), 0x2b)
 #define AVD_OP_EXEC_FIFO_MASK(v)	FIELD_PREP(GENMASK(3, 0), v)
 #define AVD_OP_EXEC_FIFO_IDX(v)		FIELD_PREP(GENMASK(7, 4), v)
diff --git a/drivers/media/platform/apple/avd/avd-vp9.c b/drivers/media/platform/apple/avd/avd-vp9.c
index d43d19e631ec4b..be0e7d021f36f4 100644
--- a/drivers/media/platform/apple/avd/avd-vp9.c
+++ b/drivers/media/platform/apple/avd/avd-vp9.c
@@ -185,9 +185,9 @@ static void set_refs(struct avd_ctx *ctx, struct avd_vp9_run *run)
 		avd_get_ref_buf(ctx, &dst->base.vb, frame->golden_frame_ts);
 	ref_buf[2] = avd_get_ref_buf(ctx, &dst->base.vb, frame->alt_frame_ts);
 
-	push(INST_DMA3, "cm3_dma_config_7");
-	push(INST_DMA3, "cm3_dma_config_8");
-	push(INST_DMA3, "cm3_dma_config_9");
+	push(0, "");
+	push(0, "");
+	push(0, "");
 
 	for (int i = 0; i < V4L2_VP9_NUM_FRAME_CTX - 1; i++) {
 		addr = vb2_dma_contig_plane_dma_addr(
@@ -306,7 +306,7 @@ static void set_header(struct avd_ctx *ctx, struct avd_vp9_run *run)
 	push(AVD_HDR_HEIGHT(frame->frame_height_minus_1) |
 		     AVD_HDR_WIDTH(frame->frame_width_minus_1),
 	     "hdr_28_height_width_shift3");
-	push(0, "cm3_dma_config_0");
+	push(0, "");
 	push(AVD_HDR_HEIGHT(frame->frame_height_minus_1) |
 		     AVD_HDR_WIDTH(frame->frame_width_minus_1),
 	     "hdr_38_height_width_shift3");
@@ -327,8 +327,8 @@ static void set_header(struct avd_ctx *ctx, struct avd_vp9_run *run)
 
 	/* some kind of feature enable? h26{2,5} has 3 instread */
 	push(AVD_HDR_FEAT_VP9, "unk_const");
-	push(INST_DMA2, "cm3_dma_config_2");
-	push(INST_DMA1, "cm3_dma_config_3");
+	push(0, "");
+	push(0, "");
 
 	pusha(vp9_ctx->bufs.counts.addr, "frame_counts_addr", 0);
 	pusha(vp9_ctx->bufs.probs.addr, "hdr_104_probs_addr_lsb8", 0);
@@ -367,8 +367,8 @@ static void set_header(struct avd_ctx *ctx, struct avd_vp9_run *run)
 			      0),
 	     "hdr_48_loop_filter_level");
 
-	push(INST_DMA2, "cm3_dma_config_4");
-	push(INST_DMA2, "cm3_dma_config_5");
+	push(0, "");
+	push(0, "");
 
 	if (avd->variant->revision == 3)
 		push(0, "");
@@ -386,7 +386,7 @@ static void set_header(struct avd_ctx *ctx, struct avd_vp9_run *run)
 	/* fatal if missing / wrong */
 	pusha(vp9_ctx->bufs.tiles[2].addr, "hdr_e8_sps0_tile_addr_lsb8", 0);
 
-	push(INST_DMA3, "cm3_dma_config_7");
+	push(0, "");
 
 	push_rvra(avd, ctx, run->addresses.rvra, ctx->rvra.offsets);
 
