diff --git a/drivers/media/platform/apple/avd/avd-drv.c b/drivers/media/platform/apple/avd/avd-drv.c
index 773de35ce883a0..efc7aa86db1868 100644
--- a/drivers/media/platform/apple/avd/avd-drv.c
+++ b/drivers/media/platform/apple/avd/avd-drv.c
@@ -97,6 +97,25 @@ void avd_buf_free(struct avd_dev *avd, struct avd_buf *buf)
 	memset(buf, 0, sizeof(*buf));
 }
 
+struct avd_decoded_buffer *
+get_ref_buf(struct avd_ctx *ctx, struct vb2_v4l2_buffer *dst, u64 timestamp)
+{
+	struct v4l2_m2m_ctx *m2m_ctx = ctx->fh.m2m_ctx;
+	struct vb2_queue *cap_q = &m2m_ctx->cap_q_ctx.q;
+	struct vb2_buffer *buf;
+
+	/*
+	 * If a ref is unused or invalid, address of current destination
+	 * buffer is returned.
+	 */
+	buf = vb2_find_buffer(cap_q, timestamp);
+	if (!buf)
+		buf = &dst->vb2_buf;
+
+	return vb2_to_avd_decoded_buf(buf);
+}
+
+
 static int avd_reset(struct avd_dev *avd)
 {
 	int ret = 0;
diff --git a/drivers/media/platform/apple/avd/avd-vp9.c b/drivers/media/platform/apple/avd/avd-vp9.c
index dbc07aafa09008..8877511db78e7b 100644
--- a/drivers/media/platform/apple/avd/avd-vp9.c
+++ b/drivers/media/platform/apple/avd/avd-vp9.c
@@ -154,24 +154,6 @@ struct avd_vp9_ctx {
 	u8 submit_num;
 };
 
-static struct avd_decoded_buffer *
-get_ref_buf(struct avd_ctx *ctx, struct vb2_v4l2_buffer *dst, u64 timestamp)
-{
-	struct v4l2_m2m_ctx *m2m_ctx = ctx->fh.m2m_ctx;
-	struct vb2_queue *cap_q = &m2m_ctx->cap_q_ctx.q;
-	struct vb2_buffer *buf;
-
-	/*
-	 * If a ref is unused or invalid, address of current destination
-	 * buffer is returned.
-	 */
-	buf = vb2_find_buffer(cap_q, timestamp);
-	if (!buf)
-		buf = &dst->vb2_buf;
-
-	return vb2_to_avd_decoded_buf(buf);
-}
-
 static void set_refs(struct avd_ctx *ctx, struct avd_vp9_run *run)
 {
 	const struct v4l2_ctrl_vp9_frame *frame = run->decode_params;
diff --git a/drivers/media/platform/apple/avd/avd.h b/drivers/media/platform/apple/avd/avd.h
index a736eaa2cc5213..88e7c77cf880dd 100644
--- a/drivers/media/platform/apple/avd/avd.h
+++ b/drivers/media/platform/apple/avd/avd.h
@@ -93,6 +93,8 @@ vb2_to_avd_decoded_buf(struct vb2_buffer *buf)
 {
 	return container_of(buf, struct avd_decoded_buffer, base.vb.vb2_buf);
 }
+struct avd_decoded_buffer *
+get_ref_buf(struct avd_ctx *ctx, struct vb2_v4l2_buffer *dst, u64 timestamp);
 
 struct avd_coded_fmt_ops {
 	void (*adjust_decoded_fmt)(struct avd_ctx *ctx,
