diff --git a/drivers/media/platform/apple/avd/avd-av1.c b/drivers/media/platform/apple/avd/avd-av1.c
index ffc532a549e9c9..fe114d06bc3c92 100644
--- a/drivers/media/platform/apple/avd/avd-av1.c
+++ b/drivers/media/platform/apple/avd/avd-av1.c
@@ -185,7 +185,7 @@
 #define AVD_CDFS_SIZE	(sizeof(struct avd_av1_cdfs))
 
 #define AVD_AV1_TLB_OFFSET(dst, tlb) \
-	(ALIGN_DOWN(dst, AVD_ALIGN) - ALIGN(tlb, AVD_ALIGN))
+	((dst) - ALIGN(tlb, AVD_ALIGN))
 #define AVD_AV1_CDFS_OFFSET(dst, tlb) \
 	(AVD_AV1_TLB_OFFSET(dst, tlb) - ALIGN(AVD_CDFS_SIZE, AVD_ALIGN))
 
@@ -193,7 +193,6 @@ struct avd_av1_run {
 	struct avd_run base;
 
 	struct {
-		dma_addr_t rvra;
 		dma_addr_t probs_out;
 		dma_addr_t priv_tlb;
 	} addresses;
@@ -341,8 +340,7 @@ static void set_refs(struct avd_ctx *ctx, struct avd_av1_run *run)
 					[frame->ref_frame_idx[i]]);
 		}
 		addr = vb2_dma_contig_plane_dma_addr(&ref->base.vb.vb2_buf, 0) +
-		       AVD_AV1_RVRA_OFFSET(ref->base.vb.planes[0].length,
-					   ref->rvra.size);
+		       ref->comp.start_offset;
 
 		if (!intrabc) {
 			int shift =
@@ -400,7 +398,7 @@ static void set_refs(struct avd_ctx *ctx, struct avd_av1_run *run)
 		push(AV1_REF_SCALE_X(x_scale) | AV1_REF_SCALE_Y(y_scale),
 		     "ref_scale");
 
-		push_rvra(avd, ctx, addr, ref->rvra.offsets);
+		push_comp(avd, ctx, addr, ref->comp.offsets);
 	}
 }
 
@@ -615,7 +613,7 @@ static void set_header(struct avd_ctx *ctx, struct avd_av1_run *run)
 		     AVD_OP_EXEC_FLAG_START_REV4(avd->variant->revision == 4) |
 		     AVD_OP_EXEC_FIFO_IDX(ctx->fifo_idx),
 	     "vp_start");
-	push(AVD_OP_HDR | AVD_OP_HDR_FLAG0 |
+	push(AVD_OP_HDR | AVD_OP_HDR_FLAG_DECOMP(ctx->decomp) |
 		     AVD_OP_HDR_FLAG_INTRA(intra_only && !intrabc) |
 		     AVD_OP_HDR_CONST | AVD_OP_HDR_FLAG_PIPE_STATE(1),
 	     "op_hdr");
@@ -918,11 +916,12 @@ static void set_header(struct avd_ctx *ctx, struct avd_av1_run *run)
 
 	push(0, "mark_section");
 
-	push_rvra(avd, ctx, run->addresses.rvra, ctx->rvra.offsets);
+	push_comp(avd, ctx, run->base.comp_out, ctx->comp.offsets);
 
 	push(0, "");
 	push(0, "mark_section");
 
+	/* ignored if decomp is disabled */
 	bytesperline = ctx->decoded_fmt.fmt.pix_mp.plane_fmt[0].bytesperline;
 	pusha(run->base.y_out, "y", 0);
 	push(bytesperline, "bytesperline_y");
@@ -1115,9 +1114,6 @@ static int avd_av1_run_preamble(struct avd_ctx *ctx, struct avd_av1_run *run)
 	tlb_len = avd_priv_tlb_size(run->frame->frame_width_minus_1 + 1,
 				    run->frame->frame_height_minus_1 + 1);
 
-	run->addresses.rvra =
-		run->addresses.y + AVD_AV1_RVRA_OFFSET(dst_len, ctx->rvra.size);
-
 	run->addresses.priv_tlb =
 		run->base.y_out + AVD_AV1_TLB_OFFSET(dst_len, tlb_len);
 
diff --git a/drivers/media/platform/apple/avd/avd-drv.c b/drivers/media/platform/apple/avd/avd-drv.c
index fc48f635bf335f..70ab0f4438a931 100644
--- a/drivers/media/platform/apple/avd/avd-drv.c
+++ b/drivers/media/platform/apple/avd/avd-drv.c
@@ -21,41 +21,50 @@
 #include "avd.h"
 #include "avd-regs.h"
 
-void fill_rvra(struct avd_rvra *rvra, enum avd_image_fmt image_fmt,
-		u32 width, u32 height)
+static void calc_tile_meta(u32 w, u32 h, u32 bpb, u32 tile_dim,
+			   u32 meta_hdr_bytes, u32 *tile, u32 *meta)
 {
-	u32 size0, size1, size2;
-	u32 hs = round_up(height, 32);
+	u32 tiles_width, tiles_height, meta_tile_w, meta_tile_h, tile_bytes;
+	tiles_width = DIV_ROUND_UP(w, tile_dim);
+	tiles_height = DIV_ROUND_UP(h, tile_dim);
+	tile_bytes = tile_dim * tile_dim * DIV_ROUND_UP(bpb, 8);
+	*tile = ALIGN(tiles_width * tiles_height * tile_bytes, 16);
 
-	size0 = (width * hs) + ((width * hs) / 4);
-	size2 = size0;
+	meta_tile_w = roundup_pow_of_two(tiles_width);
+	meta_tile_h = roundup_pow_of_two(tiles_height);
+
+	*meta = ALIGN(meta_tile_w * meta_tile_h * meta_hdr_bytes, 16);
+}
+
+void fill_comp(struct avd_comp *comp, enum avd_image_fmt image_fmt, u32 width,
+	       u32 height)
+{
+	u32 y_meta, y, uv_meta, uv;
+	int bit_depth;
 
 	switch (image_fmt) {
-	case AVD_IMG_FMT_420_8BIT:
 	case AVD_IMG_FMT_420_10BIT:
-		size2 /= 2;
+	case AVD_IMG_FMT_422_10BIT:
+		bit_depth = 10;
 		break;
 	default:
+		bit_depth = 8;
 		break;
 	}
 
-	size1 = max((roundup_pow_of_two(width) * roundup_pow_of_two(height)) / 32,
-			 0x100u);
-	/* TODO: how big? */
-
-	rvra->size = round_up(size0 + size1 + size2, 0x4000);
-	rvra->size +=
-		(width < 1000 ? 0 : width < 1800 ? 2 : width < 3800 ? 3 : 9) * 0x4000;
-	/* TODO */
-	rvra->size +=
-		(height < 1000 ? 0 : height < 1800 ? 2 : height < 3800 ? 3 : 9) * 0x4000;
-	/* TODO */
-	rvra->size += 0x10000;
-
-	rvra->offsets[1] = 0;
-	rvra->offsets[0] = size0;
-	rvra->offsets[3] = size0 + size1;
-	rvra->offsets[2] = size0 + size1 + size2;
+	/* y has 32x32 tiles and 32 bytes of metadata per tile */
+	calc_tile_meta(width, height, bit_depth, 32, 32, &y, &y_meta);
+	/* uv has 16x16 tiles and 8 bytes of metadata per tile */
+	calc_tile_meta(width / 2, height / 2, bit_depth * 2, 16, 8, &uv,
+		       &uv_meta);
+
+	/* output like DCP driver expects */
+	comp->offsets[0] = y;
+	comp->offsets[1] = 0;
+	comp->offsets[2] = y + y_meta + uv;
+	comp->offsets[3] = y + y_meta;
+
+	comp->size = y_meta + y + uv_meta + uv;
 }
 
 
diff --git a/drivers/media/platform/apple/avd/avd-h264.c b/drivers/media/platform/apple/avd/avd-h264.c
index 2a63ebbbddfcac..1c77ed44ad5ab0 100644
--- a/drivers/media/platform/apple/avd/avd-h264.c
+++ b/drivers/media/platform/apple/avd/avd-h264.c
@@ -43,7 +43,6 @@ struct avd_h264_run {
 	const struct v4l2_ctrl_h264_pred_weights *pred_weights;
 
 	struct run_addr {
-		dma_addr_t rvra;
 		dma_addr_t sps;
 	} addresses;
 
@@ -96,6 +95,10 @@ static void stream_refs(struct avd_ctx *ctx, struct avd_h264_run *run)
 	const struct v4l2_h264_dpb_entry *dpb = decode->dpb;
 	struct avd_h264_ctx *h264_ctx = ctx->priv;
 	struct avd_dev *avd = ctx->dev;
+	struct avd_decoded_buffer *dst, *ref;
+	dma_addr_t addr;
+
+	dst = vb2_to_avd_decoded_buf(&run->base.bufs.dst->vb2_buf);
 
 	push(0, "");
 	pusha(h264_ctx->bufs.pps_tile[4].addr, "hdr_9c_pps_tile_addr_lsb8", 7);
@@ -110,16 +113,10 @@ static void stream_refs(struct avd_ctx *ctx, struct avd_h264_run *run)
 		if (!(dpb[i].flags & V4L2_H264_DPB_ENTRY_FLAG_VALID))
 			continue;
 
-		struct vb2_buffer *vb = vb2_find_buffer(
-			&ctx->fh.m2m_ctx->cap_q_ctx.q, dpb[i].reference_ts);
+		ref = avd_get_ref_buf(ctx, &dst->base.vb, dpb[i].reference_ts);
 
-		dma_addr_t rvra_addr =
-			vb ? vb2_dma_contig_plane_dma_addr(vb, 0) +
-					(vb->planes[0].length -
-					 sps_size(fmt_width(ctx),
-						  fmt_height(ctx)) -
-					 ctx->rvra.size) :
-			     run->addresses.rvra; /* safe fallback */
+		addr = vb2_dma_contig_plane_dma_addr(&ref->base.vb.vb2_buf, 0) +
+		       ref->comp.start_offset;
 
 		push(AVD_REF_NUM(run->num_valid - 1) | AVD_REF_FLAG_CONST |
 			     AVD_REF_FLAG_LONG(
@@ -129,7 +126,7 @@ static void stream_refs(struct avd_ctx *ctx, struct avd_h264_run *run)
 					       dpb[i].top_field_order_cnt),
 		     "hdr_d0_ref_hdr");
 
-		push_rvra(avd, ctx, rvra_addr, ctx->rvra.offsets);
+		push_comp(avd, ctx, addr, ctx->comp.offsets);
 	}
 }
 
@@ -205,7 +202,7 @@ static void stream_hdr(struct avd_ctx *ctx, struct avd_h264_run *run)
 		     AVD_OP_EXEC_FLAG_START_REV4(avd->variant->revision == 4),
 	     "inst_fifo_start");
 
-	push(AVD_OP_HDR | AVD_OP_HDR_FLAG0 |
+	push(AVD_OP_HDR | AVD_OP_HDR_FLAG_DECOMP(ctx->decomp) |
 		     AVD_OP_HDR_FLAG_INTRA(
 			     decode->flags &
 			     V4L2_H264_DECODE_PARAM_FLAG_IDR_PIC) |
@@ -279,7 +276,7 @@ static void stream_hdr(struct avd_ctx *ctx, struct avd_h264_run *run)
 	pusha(h264_ctx->bufs.pps_tile[3].addr, "hdr_9c_pps_tile_addr_lsb8", 3);
 	push(0, "");
 
-	push_rvra(avd, ctx, run->addresses.rvra, ctx->rvra.offsets);
+	push_comp(avd, ctx, run->base.comp_out, ctx->comp.offsets);
 
 	bytesperline = ctx->decoded_fmt.fmt.pix_mp.plane_fmt[0].bytesperline;
 	if (avd->variant->quirks & AVD_QUIRK_LSR)
@@ -677,9 +674,6 @@ static void avd_h264_run_preamble(struct avd_ctx *ctx, struct avd_h264_run *run)
 
 	sps_len = sps_size(fmt_width(ctx), fmt_height(ctx));
 
-	run->addresses.rvra =
-		run->addresses.y + (dst_len - sps_len - ctx->rvra.size);
-
 	run->addresses.sps = run->base.y_out + (dst_len - sps_len);
 }
 
diff --git a/drivers/media/platform/apple/avd/avd-hevc.c b/drivers/media/platform/apple/avd/avd-hevc.c
index 1faddc0a24c25a..ec40586d14582a 100644
--- a/drivers/media/platform/apple/avd/avd-hevc.c
+++ b/drivers/media/platform/apple/avd/avd-hevc.c
@@ -71,7 +71,6 @@ struct avd_hevc_run {
 	int num_slices;
 
 	struct run_addr {
-		dma_addr_t rvra;
 		dma_addr_t sps;
 	} addresses;
 };
@@ -114,12 +113,9 @@ static void stream_refs(struct avd_ctx *ctx, struct avd_hevc_run *run)
 
 		ref_buf = avd_get_ref_buf(ctx, &dst->base.vb, dpb->timestamp);
 
-		dma_addr_t rvra_addr =
-			vb2_dma_contig_plane_dma_addr(&ref_buf->base.vb.vb2_buf,
-						      0) +
-			(ref_buf->base.vb.planes[0].length -
-			 sps_size(fmt_width(ctx), fmt_height(ctx)) -
-			 ref_buf->rvra.size);
+		dma_addr_t comp_addr = vb2_dma_contig_plane_dma_addr(
+					       &ref_buf->base.vb.vb2_buf, 0) +
+				       ref_buf->comp.start_offset;
 
 		push(AVD_REF_NUM(decode->num_active_dpb_entries - 1) |
 			     AVD_REF_FLAG_CONST |
@@ -130,7 +126,7 @@ static void stream_refs(struct avd_ctx *ctx, struct avd_hevc_run *run)
 					       dpb->pic_order_cnt_val),
 		     "hdr_d0_ref_hdr");
 
-		push_rvra(avd, ctx, rvra_addr, ref_buf->rvra.offsets);
+		push_comp(avd, ctx, comp_addr, ref_buf->comp.offsets);
 	}
 }
 
@@ -301,8 +297,8 @@ static void set_header(struct avd_ctx *ctx, struct avd_hevc_run *run)
 		     AVD_OP_EXEC_FLAG_START_REV4(avd->variant->revision == 4),
 	     "inst_fifo_start");
 
-	push(AVD_OP_HDR | AVD_OP_HDR_FLAG0 | AVD_OP_HDR_FLAG_INTRA(is_intra) |
-		     AVD_OP_HDR_CONST |
+	push(AVD_OP_HDR | AVD_OP_HDR_FLAG_DECOMP(ctx->decomp) |
+		     AVD_OP_HDR_FLAG_INTRA(is_intra) | AVD_OP_HDR_CONST |
 		     AVD_OP_HDR_FLAG_PIPE_STATE(
 			     !(avd->variant->quirks & AVD_QUIRK_NO_PIPE_STATE)),
 	     "hdr_34_start_hdr");
@@ -381,7 +377,7 @@ static void set_header(struct avd_ctx *ctx, struct avd_hevc_run *run)
 
 	push(0, "");
 
-	push_rvra(avd, ctx, run->addresses.rvra, ctx->rvra.offsets);
+	push_comp(avd, ctx, run->base.comp_out, ctx->comp.offsets);
 
 	push(0, "cm3_mark_end_section");
 
@@ -1221,9 +1217,6 @@ static int avd_hevc_run_preamble(struct avd_ctx *ctx, struct avd_hevc_run *run)
 
 	sps_len = sps_size(fmt_width(ctx), fmt_height(ctx));
 
-	run->addresses.rvra =
-		run->addresses.y + (dst_len - sps_len - ctx->rvra.size);
-
 	run->addresses.sps = run->base.y_out + (dst_len - sps_len);
 	return 0;
 }
diff --git a/drivers/media/platform/apple/avd/avd-inst.h b/drivers/media/platform/apple/avd/avd-inst.h
index 41815339b8a06f..db741db572e77c 100644
--- a/drivers/media/platform/apple/avd/avd-inst.h
+++ b/drivers/media/platform/apple/avd/avd-inst.h
@@ -17,7 +17,8 @@
 
 #define AVD_OP_HDR			FIELD_PREP(GENMASK(31, 20), 0x2db)
 #define AVD_OP_HDR_CONST		FIELD_PREP(GENMASK(10, 0), 0x2e0)
-#define AVD_OP_HDR_FLAG0		FIELD_PREP(BIT(12), 1)
+/* decompress pixel data */
+#define AVD_OP_HDR_FLAG_DECOMP(v)	FIELD_PREP(BIT(12), !!(v))
 #define AVD_OP_HDR_FLAG_INTRA(v)	FIELD_PREP(BIT(13), !!(v))
 #define AVD_OP_HDR_FLAG_PIPE_STATE(v)	FIELD_PREP(BIT(19), !!(v))
 
@@ -167,7 +168,7 @@ static inline void push_address(struct avd_dev *avd, struct avd_ctx *ctx,
 		push(avd, ctx, (u32)(addr >> 32));
 	}
 }
-static inline void push_rvra(struct avd_dev *avd, struct avd_ctx *ctx,
+static inline void push_comp(struct avd_dev *avd, struct avd_ctx *ctx,
 		dma_addr_t addr, u32 offsets[4])
 {
 	if (avd->variant->quirks & AVD_QUIRK_LSR) {
diff --git a/drivers/media/platform/apple/avd/avd-v4l2.c b/drivers/media/platform/apple/avd/avd-v4l2.c
index 4cd7606cb352ac..7d4b3e138cf994 100644
--- a/drivers/media/platform/apple/avd/avd-v4l2.c
+++ b/drivers/media/platform/apple/avd/avd-v4l2.c
@@ -78,8 +78,10 @@ static void avd_fill_decoded_pixfmt(struct avd_ctx *ctx,
 {
 	v4l2_fill_pixfmt_mp(pix_mp, pix_mp->pixelformat, pix_mp->width,
 			    pix_mp->height);
-	fill_rvra(&ctx->rvra, ctx->image_fmt, pix_mp->width, pix_mp->height);
-	pix_mp->plane_fmt[0].sizeimage += ctx->rvra.size;
+
+	ctx->comp.start_offset = pix_mp->plane_fmt[0].sizeimage;
+	fill_comp(&ctx->comp, ctx->image_fmt, pix_mp->width, pix_mp->height);
+	pix_mp->plane_fmt[0].sizeimage += ctx->comp.size;
 	if (ctx->coded_fmt_desc->ops->adjust_decoded_fmt)
 		ctx->coded_fmt_desc->ops->adjust_decoded_fmt(ctx, pix_mp);
 }
@@ -960,8 +962,11 @@ void avd_run_preamble(struct avd_ctx *ctx, struct avd_run *run)
 		      ctx->decoded_fmt.fmt.pix_mp.plane_fmt[0].bytesperline *
 			      ctx->decoded_fmt.fmt.pix_mp.height;
 
+	run->comp_out = run->y_out + ctx->comp.start_offset;
+	ctx->decomp = true;
+
 	dst = vb2_to_avd_decoded_buf(&run->bufs.dst->vb2_buf);
-	memcpy(&dst->rvra, &ctx->rvra, sizeof(ctx->rvra));
+	memcpy(&dst->comp, &ctx->comp, sizeof(ctx->comp));
 
 	/* Apply request(s) controls if needed. */
 	src_req = run->bufs.src->vb2_buf.req_obj.req;
diff --git a/drivers/media/platform/apple/avd/avd-vp9.c b/drivers/media/platform/apple/avd/avd-vp9.c
index 4f0c066d4477ac..f331b9cf4057a5 100644
--- a/drivers/media/platform/apple/avd/avd-vp9.c
+++ b/drivers/media/platform/apple/avd/avd-vp9.c
@@ -137,10 +137,6 @@ struct avd_vp9_frame_info {
 struct avd_vp9_run {
 	struct avd_run base;
 
-	struct run_addr {
-		dma_addr_t rvra;
-	} addresses;
-
 	const struct v4l2_ctrl_vp9_frame *decode_params;
 	const struct v4l2_ctrl_vp9_compressed_hdr *prob_updates;
 };
@@ -189,8 +185,7 @@ static void set_refs(struct avd_ctx *ctx, struct avd_vp9_run *run)
 	for (int i = 0; i < V4L2_VP9_NUM_FRAME_CTX - 1; i++) {
 		addr = vb2_dma_contig_plane_dma_addr(
 			       &ref_buf[i]->base.vb.vb2_buf, 0) +
-		       (ref_buf[i]->base.vb.planes[0].length -
-			ref_buf[i]->rvra.size);
+		       ref_buf[i]->comp.start_offset;
 
 		/* TODO */
 		push(AVD_REF_FLAG_CONST, "hdr_9c_ref_100");
@@ -199,7 +194,7 @@ static void set_refs(struct avd_ctx *ctx, struct avd_vp9_run *run)
 		     "hdr_70_ref_height_width");
 		push(0x40004000, "hdr_7c_ref_align");
 
-		push_rvra(avd, ctx, addr, ref_buf[i]->rvra.offsets);
+		push_comp(avd, ctx, addr, ref_buf[i]->comp.offsets);
 	}
 }
 
@@ -292,8 +287,8 @@ static void set_header(struct avd_ctx *ctx, struct avd_vp9_run *run)
 		     AVD_OP_EXEC_FIFO_IDX(ctx->fifo_idx),
 	     "inst_fifo_start");
 
-	push(AVD_OP_HDR | AVD_OP_HDR_FLAG0 | AVD_OP_HDR_FLAG_INTRA(intra_only) |
-		     AVD_OP_HDR_CONST |
+	push(AVD_OP_HDR | AVD_OP_HDR_FLAG_DECOMP(ctx->decomp) |
+		     AVD_OP_HDR_FLAG_INTRA(intra_only) | AVD_OP_HDR_CONST |
 		     AVD_OP_HDR_FLAG_PIPE_STATE(
 			     !(avd->variant->quirks & AVD_QUIRK_NO_PIPE_STATE)),
 	     "hdr_34_start_hdr");
@@ -385,7 +380,7 @@ static void set_header(struct avd_ctx *ctx, struct avd_vp9_run *run)
 
 	push(0, "");
 
-	push_rvra(avd, ctx, run->addresses.rvra, ctx->rvra.offsets);
+	push_comp(avd, ctx, run->base.comp_out, ctx->comp.offsets);
 
 	/* confusing */
 	pusha((u64)0, "hdr_f4_sps1_tile_addr_lsb8", 2);
@@ -728,9 +723,6 @@ static int avd_vp9_run_preamble(struct avd_ctx *ctx, struct avd_vp9_run *run)
 	v4l2_vp9_fw_update_probs(&vp9_ctx->probability_tables,
 				 run->prob_updates, dec_params);
 
-
-	run->addresses.rvra = run->addresses.y + (dst_len - ctx->rvra.size);
-
 	return 0;
 }
 
diff --git a/drivers/media/platform/apple/avd/avd.h b/drivers/media/platform/apple/avd/avd.h
index a40e5c59d36cf2..92252934cb0d5a 100644
--- a/drivers/media/platform/apple/avd/avd.h
+++ b/drivers/media/platform/apple/avd/avd.h
@@ -60,6 +60,7 @@ struct avd_run {
 	dma_addr_t coded_in;
 	dma_addr_t y_out;
 	dma_addr_t uv_out;
+	dma_addr_t comp_out;
 };
 
 struct avd_ctrl_desc {
@@ -97,16 +98,19 @@ struct avd_hevc_decoded_buffer_info {
 };
 
 
-struct avd_rvra {
-	u32 offsets[4]; /* sizes or offsets */
+struct avd_comp {
 	u32 size;
+	/* offset to start of compressed data */
+	size_t start_offset;
+	/* relative offsets to start */
+	u32 offsets[4];
 };
 
 struct avd_decoded_buffer {
 	/* Must be the first field in this struct. */
 	struct v4l2_m2m_buffer base;
 
-	struct avd_rvra rvra;
+	struct avd_comp comp;
 
 	union {
 		struct avd_vp9_decoded_buffer_info vp9;
@@ -217,13 +221,13 @@ struct avd_ctx {
 	const struct avd_coded_fmt_desc *coded_fmt_desc;
 	struct v4l2_ctrl_handler ctrl_hdl;
 	enum avd_image_fmt image_fmt;
+	bool decomp;
 
 	struct delayed_work watchdog_work;
 
 	void *priv;
 
-	/* reference VRA (video resolution adaptation) scaler buffer. */
-	struct avd_rvra rvra;
+	struct avd_comp comp;
 
 	u8 fifo_idx;
 	u8 vp_slot;
@@ -266,7 +270,7 @@ static inline u32 fmt_width(struct avd_ctx *ctx)
 	return ctx->coded_fmt.fmt.pix_mp.width;
 }
 
-void fill_rvra(struct avd_rvra *rvra, enum avd_image_fmt image_fmt,
+void fill_comp(struct avd_comp *comp, enum avd_image_fmt image_fmt,
 		u32 width, u32 height);
 int alloc_slots(struct avd_dev *avd, struct avd_ctx *ctx, enum avd_codec codec);
 
