diff --git a/drivers/media/platform/apple/avd/avd-vp9.c b/drivers/media/platform/apple/avd/avd-vp9.c
index f331b9cf4057a5..869c7257c5960e 100644
--- a/drivers/media/platform/apple/avd/avd-vp9.c
+++ b/drivers/media/platform/apple/avd/avd-vp9.c
@@ -41,6 +41,8 @@
 
 #define VP9_FEAT_LVL_ALT_Q(v)	FIELD_PREP(GENMASK(21, 12), v)
 
+#define VP9_MAX_TILE_COLS	(1 << 4)
+
 struct avd_vp9_seg_probs {
 	u8 tree_probs[7];
 	u8 pred_probs[3];
@@ -612,7 +614,13 @@ static int avd_vp9_alloc_bufs(struct avd_ctx *ctx)
 {
 	struct avd_dev *avd = ctx->dev;
 	struct avd_vp9_ctx *vp9_ctx = ctx->priv;
-	int ret, size;
+	int ret, w, h, bit_depth;
+	w = fmt_width(ctx);
+	h = fmt_height(ctx);
+	bit_depth = (ctx->image_fmt == AVD_IMG_FMT_420_10BIT ||
+		     ctx->image_fmt == AVD_IMG_FMT_422_10BIT) ?
+			    10 :
+			    8;
 
 	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.inst, fifo_size());
 	if (ret)
@@ -624,66 +632,61 @@ static int avd_vp9_alloc_bufs(struct avd_ctx *ctx)
 			return ret;
 	}
 
-	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.seg, 0x200);
+	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.probs,
+			    sizeof(struct avd_vp9_probs));
+	if (ret)
+		return ret;
+
+	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.counts,
+			    sizeof(struct avd_vp9_frame_symbol_counts));
 	if (ret)
 		return ret;
 
-	/* scratch buffer? */
-	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.tiles[1], 0x200);
+	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.above_info,
+			    DIV_ROUND_UP(w, 16) * DIV_ROUND_UP(h, 64) * 144);
 	if (ret)
 		return ret;
 
-	/* TODO */
-	size = (((ctx->decoded_fmt.fmt.pix_mp.height - 1) *
-		 (ctx->decoded_fmt.fmt.pix_mp.height - 1) / 0x10000) +
-		2) *
-	       0x4000;
-	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.tiles[0], size);
+	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.color[0],
+			    DIV_ROUND_UP(w, 16) * 4 * bit_depth +
+				    (VP9_MAX_TILE_COLS - 1) * 128);
 	if (ret)
 		return ret;
-	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.tiles[2], size);
+	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.color[1],
+			    DIV_ROUND_UP(w, 8) * 16 * bit_depth);
 	if (ret)
 		return ret;
 
-	for (int i = 0; i < 2; i++) {
-		ret = avd_buf_alloc(avd, &vp9_ctx->bufs.color[i], size);
-		if (ret)
-			return ret;
-	}
+	/*
+	 * randomly needs more space when resizing?
+	 * maybe it does not need it?
+	 */
+	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.seg, DIV_ROUND_UP(w, 8) * 24);
+	if (ret)
+		return ret;
 
-	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.above_info, size + 0xc000);
+	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.tiles[0],
+			    DIV_ROUND_UP(h, 8) * 16 * bit_depth);
 	if (ret)
 		return ret;
 
-	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.state, size);
+	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.tiles[1],
+			    DIV_ROUND_UP(h, 64) * 16);
 	if (ret)
 		return ret;
 
-	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.probs,
-			    sizeof(struct avd_vp9_probs));
+	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.tiles[2],
+			    bit_depth * 36 * DIV_ROUND_UP(h, 8) +
+				    bit_depth * 18 * DIV_ROUND_UP(h, 16));
 	if (ret)
 		return ret;
 
-	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.counts,
-			    sizeof(struct avd_vp9_frame_symbol_counts));
+	ret = avd_buf_alloc(avd, &vp9_ctx->bufs.state,
+			    DIV_ROUND_UP(w, 64) * 288 +
+				    (VP9_MAX_TILE_COLS - 1) * 128);
 	if (ret)
 		return ret;
 
-#ifdef AVD_DEBUG
-#define DBG_BUF(name, buf)                                           \
-	pr_info("   %10s %12llx %8lx", name, vp9_ctx->bufs.buf.addr, \
-		vp9_ctx->bufs.buf.size)
-	pr_info("buffers:");
-	DBG_BUF("state", state);
-	DBG_BUF("above_info", above_info);
-	DBG_BUF("seg", seg);
-	DBG_BUF("tiles[0]", tiles[0]);
-	DBG_BUF("tiles[1]", tiles[1]);
-	DBG_BUF("tiles[2]", tiles[2]);
-	DBG_BUF("color[0]", color[0]);
-	DBG_BUF("color[1]", color[1]);
-	pr_info("\n");
-#endif
 	return 0;
 }
 
