diff --git a/drivers/media/platform/apple/avd/avd-h264.c b/drivers/media/platform/apple/avd/avd-h264.c
index c6688a5fec75af..7309151aab1d91 100644
--- a/drivers/media/platform/apple/avd/avd-h264.c
+++ b/drivers/media/platform/apple/avd/avd-h264.c
@@ -29,6 +29,8 @@
 #define H264_FLAG_ENTROPY_CODING_MODE(v)	FIELD_PREP(BIT(20), !!(v))
 #define H264_FLAG_NOT_IDR(v)			FIELD_PREP(BIT(21), !!(v))
 
+#define H264_TRANSFORM_8X8_MODE(v)		FIELD_PREP(BIT(7), !!(v))
+
 struct avd_h264_run {
 	struct avd_run base;
 
@@ -232,7 +234,7 @@ static void stream_hdr(struct avd_ctx *ctx, struct avd_h264_run *run)
 		     AVD_HDR_COMMON_BIT_DEPTH_C(sps->bit_depth_chroma_minus8) |
 		     AVD_HDR_COMMON_MIN_LUMA_CBS(1) |
 		     AVD_HDR_COMMON_LUMA_CBS(1) |
-		     AVD_HDR_COMMON_LUMA_TBS(
+		     H264_TRANSFORM_8X8_MODE(
 			     pps->flags &
 			     V4L2_H264_PPS_FLAG_TRANSFORM_8X8_MODE) |
 		     AVD_HDR_COMMON_FLAG0(
diff --git a/drivers/media/platform/apple/avd/avd-vp9.c b/drivers/media/platform/apple/avd/avd-vp9.c
index 277d84b619b2c3..d43d19e631ec4b 100644
--- a/drivers/media/platform/apple/avd/avd-vp9.c
+++ b/drivers/media/platform/apple/avd/avd-vp9.c
@@ -805,7 +805,6 @@ static void avd_vp9_done(struct avd_ctx *ctx, struct vb2_v4l2_buffer *src_buf,
 {
 	struct avd_vp9_ctx *vp9_ctx = ctx->priv;
 	unsigned int fctx_idx;
-	avd_status(ctx->dev, 8);
 
 	/* v4l2-specific stuff */
 	if (result == VB2_BUF_STATE_ERROR)
