diff --git a/drivers/media/platform/apple/avd/avd-h264.c b/drivers/media/platform/apple/avd/avd-h264.c
index a3cfc5dd34e5d6..bfa43481441fac 100644
--- a/drivers/media/platform/apple/avd/avd-h264.c
+++ b/drivers/media/platform/apple/avd/avd-h264.c
@@ -23,6 +23,13 @@
 #include "avd.h"
 #include "avd-inst.h"
 
+#define H264_SCL_DIMS (0x1000000 | ((64 / 4) << 5) | (((16 / 4) << 5) - 1))
+
+#define h264_FLAG_CI_PRED(v)			FIELD_PREP(BIT(19), !!(v))
+#define H264_FLAG_ENTROPY_CODING_MODE(v)	FIELD_PREP(BIT(20), !!(v))
+#define H264_FLAG_NOT_IDR(v)			FIELD_PREP(BIT(21), !!(v))
+
+
 struct avd_h264_run {
 	struct avd_run base;
 
@@ -117,10 +124,10 @@ static void stream_refs(struct avd_ctx *ctx, struct avd_h264_run *run)
 						- ctx->rvra.size)
 			: run->addresses.rvra; /* safe fallback */
 
-		push(((run->num_valid - 1) & 0xf) << 28
-				| 0x1000000
-				| (boolify(dpb[i].flags & V4L2_H264_DPB_ENTRY_FLAG_LONG_TERM) << 17)
-				| (swrap(run->cur_poc - dpb[i].top_field_order_cnt, 1 << 17)),
+		push(AVD_REF_NUM(run->num_valid - 1)
+				| AVD_REF_FLAG_CONST
+				| AVD_REF_FLAG_LONG(dpb[i].flags & V4L2_H264_DPB_ENTRY_FLAG_LONG_TERM)
+				| AVD_REF_DELTA_POC(run->cur_poc - dpb[i].top_field_order_cnt),
 				"hdr_d0_ref_hdr");
 
 		push_rvra(avd, ctx, rvra_addr, ctx->rvra.offsets);
@@ -133,15 +140,14 @@ static void stream_scaling(struct avd_ctx *ctx, struct avd_h264_run *run)
 	const struct v4l2_ctrl_h264_scaling_matrix *scaling = run->scaling_matrix;
 	struct avd_dev *avd = ctx->dev;
 
-	push(0x1000000 | ((64 / 4) << 5)
-		| (((16 / 4) << 5) - 1), "hdr_4c_pic_scaling_list_dims");
+	push(H264_SCL_DIMS, "hdr_4c_pic_scaling_list_dims");
 
 	for (int i = 0; i < 6; i++)
 		for (int j = 0; j < 16; j+=4)
-			push((scaling->scaling_list_4x4[i][j + 0] << 24)
-					| (scaling->scaling_list_4x4[i][j + 1] << 16)
-					| (scaling->scaling_list_4x4[i][j + 2] << 8)
-					| (scaling->scaling_list_4x4[i][j + 3]),
+			push(AVD_SCALING_I3(scaling->scaling_list_4x4[i][j + 0])
+					| AVD_SCALING_I2(scaling->scaling_list_4x4[i][j + 1])
+					| AVD_SCALING_I1(scaling->scaling_list_4x4[i][j + 2])
+					| AVD_SCALING_I0(scaling->scaling_list_4x4[i][j + 3]),
 					"scl_46c_pic_scaling_matrix_4x4");
 
 	/* Instead of 8x8 raster scan order avd expects 4 4x4 subblocks */
@@ -156,10 +162,10 @@ static void stream_scaling(struct avd_ctx *ctx, struct avd_h264_run *run)
 	if (pps->flags & V4L2_H264_PPS_FLAG_TRANSFORM_8X8_MODE) {
 		for (int i = 0; i < 2; i++)
 			for (int j = 0; j < 16; j++)
-				push((scaling->scaling_list_8x8[i][map[j] + 0] << 24)
-						| (scaling->scaling_list_8x8[i][map[j] + 1] << 16)
-						| (scaling->scaling_list_8x8[i][map[j] + 2] << 8)
-						| (scaling->scaling_list_8x8[i][map[j] + 3]),
+				push(AVD_SCALING_I3(scaling->scaling_list_8x8[i][map[j] + 0])
+						| AVD_SCALING_I2(scaling->scaling_list_8x8[i][map[j] + 1])
+						| AVD_SCALING_I1(scaling->scaling_list_8x8[i][map[j] + 2])
+						| AVD_SCALING_I0(scaling->scaling_list_8x8[i][map[j] + 3]),
 						"scl_4cc_pic_scaling_matrix_8x8");
 
 	} else {
@@ -181,47 +187,53 @@ static void stream_hdr(struct avd_ctx *ctx, struct avd_h264_run *run)
 	u32 width = (sps->pic_width_in_mbs_minus1 + 1) * 16;
 	u32 height = (sps->pic_height_in_map_units_minus1 + 1) * 16;
 
-	push(0x2b000000
-			| (ctx->fifo_idx << 4)
-			| (avd->variant->revision == 3 ? 0x100 : 0x200),
+	push(AVD_OP_EXEC
+			| AVD_OP_EXEC_FIFO_IDX(ctx->fifo_idx)
+			| AVD_OP_EXEC_FLAG_START_REV3(avd->variant->revision == 3)
+			| AVD_OP_EXEC_FLAG_START_REV4(avd->variant->revision == 4),
 			"inst_fifo_start");
 
-	push(0x2db00000
-			| 0x1000
-			| ((decode->flags & V4L2_H264_DECODE_PARAM_FLAG_IDR_PIC) ? 0x2000 : 0)
-			| 0x2e0
-			| (avd->variant->quirks & AVD_QUIRK_NO_PIPE_STATE ? 0 : 0x80000)
+	push(AVD_OP_HDR
+			| AVD_OP_HDR_FLAG0
+			| AVD_OP_HDR_FLAG_INTRA(decode->flags & V4L2_H264_DECODE_PARAM_FLAG_IDR_PIC)
+			| AVD_OP_HDR_CONST
+			| AVD_OP_HDR_FLAG_PIPE_STATE(!(avd->variant->quirks & AVD_QUIRK_NO_PIPE_STATE))
 			, "hdr_34_start_hdr");
 
-	push(AVD_CODEC_H264 << 24, "hdr_38_mode");
+	push(AVD_HDR_CODEC_MODE(AVD_CODEC_H264), "hdr_38_mode");
 
-	push(((height - 1) << 16) | (width - 1), "hdr_3c_height_width");
+	push(AVD_HDR_HEIGHT(height - 1) | AVD_HDR_WIDTH(width - 1),
+			"hdr_3c_height_width");
 
 	push(0, "hdr_40_zero");
 
-	push((((height - 1) >> 3) << 16) | ((width - 1) >> 3),
+	push(AVD_HDR_HEIGHT((height - 1) >> 3) | AVD_HDR_WIDTH((width - 1) >> 3),
 			"hdr_28_height_width_shift3");
 
-	push((sps->chroma_format_idc & 3) << 24
-		| (sps->bit_depth_luma_minus8 & 15) << 19
-		| (sps->bit_depth_chroma_minus8 & 15) << 15
-		| 0x2800
-		| boolify(pps->flags & V4L2_H264_PPS_FLAG_TRANSFORM_8X8_MODE) << 7
-		| boolify(sps->flags & V4L2_H264_SPS_FLAG_DIRECT_8X8_INFERENCE),
+	/* TODO: did i mix up luma/chroma in hevc or here? */
+	push(AVD_HDR_COMMON_CHROMA_FORMAT(sps->chroma_format_idc)
+		| AVD_HDR_COMMON_BIT_DEPTH_L(sps->bit_depth_luma_minus8)
+		| AVD_HDR_COMMON_BIT_DEPTH_C(sps->bit_depth_chroma_minus8)
+		| AVD_HDR_COMMON_MIN_LUMA_CBS(1)
+		| AVD_HDR_COMMON_LUMA_CBS(1)
+		| AVD_HDR_COMMON_LUMA_TBS(pps->flags & V4L2_H264_PPS_FLAG_TRANSFORM_8X8_MODE)
+		| AVD_HDR_COMMON_FLAG0(sps->flags & V4L2_H264_SPS_FLAG_DIRECT_8X8_INFERENCE),
 		"hdr_2c_sps_param");
 
-	push(boolify(pps->flags & V4L2_H264_PPS_FLAG_ENTROPY_CODING_MODE) << 20
-		| !boolify(decode->flags & V4L2_H264_DECODE_PARAM_FLAG_IDR_PIC) << 21
-		| boolify(pps->flags & V4L2_H264_PPS_FLAG_CONSTRAINED_INTRA_PRED) << 19,
+	push(H264_FLAG_ENTROPY_CODING_MODE(pps->flags & V4L2_H264_PPS_FLAG_ENTROPY_CODING_MODE)
+		| H264_FLAG_NOT_IDR(!(decode->flags & V4L2_H264_DECODE_PARAM_FLAG_IDR_PIC))
+		| h264_FLAG_CI_PRED(pps->flags & V4L2_H264_PPS_FLAG_CONSTRAINED_INTRA_PRED),
 		"hdr_44_flags");
 
 
-	push(swrap(pps->chroma_qp_index_offset, 32) << 5
-		| swrap(pps->second_chroma_qp_index_offset, 32)
+	push(AVD_HDR_H26X_QP_OFFSET_CB(pps->chroma_qp_index_offset)
+		| AVD_HDR_H26X_QP_OFFSET_CR(pps->second_chroma_qp_index_offset)
 		, "hdr_48_chroma_qp_index_offset");
 
-    push(0x30000a
-			| (avd->variant->quirks & AVD_QUIRK_NO_PIPE_STATE ? 0 : 0x30),
+	push(AVD_HDR_FEAT_H26X |
+			AVD_HDR_FEAT_COMMON |
+			AVD_HDR_FEAT_H264 |
+			AVD_HDR_FEAT_PIPE_STATE_EN(!(avd->variant->quirks & AVD_QUIRK_NO_PIPE_STATE)),
 			"hdr_58_const_3a");
 
 	push(INST_DMA2, "cm3_dma_config_1");
@@ -256,7 +268,7 @@ static void stream_hdr(struct avd_ctx *ctx, struct avd_h264_run *run)
 	pusha(run->addresses.uv,"hdr_214_uv_addr_lsb8", 0);
 	push(bytesperline, "hdr_21c_width_align");
 
-	push(0x0, "cm3_mark_end_section");
+	push(0, "cm3_mark_end_section");
 	push(((height - 1) << 16) | (width - 1), "hdr_54_height_width");
 
 	if (!(decode->flags & V4L2_H264_DECODE_PARAM_FLAG_IDR_PIC))
@@ -265,12 +277,14 @@ static void stream_hdr(struct avd_ctx *ctx, struct avd_h264_run *run)
 	if (pps->flags & V4L2_H264_PPS_FLAG_SCALING_MATRIX_PRESENT)
 		stream_scaling(ctx, run);
 	else
-		push(0x0, "cm3_mark_end_section_scl");
+		push(0, "cm3_mark_end_section_scl");
 }
 
+#define DEFAULT_WEIGHT_DENOM (AVD_OP_WEIGHTS_HDR_LUMA(5) | AVD_OP_WEIGHTS_HDR_CHROMA(5))
+
 static void stream_weights(struct avd_ctx *ctx, struct avd_h264_run *run)
 {
-	int default_luma_weight, default_chroma_weight;
+	int luma_denom, chroma_denom;
 	struct v4l2_h264_weight_factors factors;
 	const struct v4l2_ctrl_h264_pred_weights *weights = run->pred_weights;
 	const struct v4l2_ctrl_h264_pps *pps = run->pps;
@@ -281,20 +295,20 @@ static void stream_weights(struct avd_ctx *ctx, struct avd_h264_run *run)
 
 	/* TODO: is there a better flag or something for the
 	 * pps->weighted_bipred_idc == 2 checks? */
-	push(0x2dd00000
-		| ((pps->weighted_bipred_idc == 2) << 7)
-		| (pred_weight << 6)
-		| (weights->luma_log2_weight_denom << 3)
-		| (weights->chroma_log2_weight_denom)
+	push(AVD_OP_WEIGHTS_HDR
+		| AVD_OP_WEIGHTS_HDR_FLAG1(pps->weighted_bipred_idc == 2)
+		| AVD_OP_WEIGHTS_HDR_FLAG0(pred_weight)
+		| AVD_OP_WEIGHTS_HDR_LUMA(weights->luma_log2_weight_denom)
+		| AVD_OP_WEIGHTS_HDR_CHROMA(weights->chroma_log2_weight_denom)
 		/* default luma and chroma denom */
-		| (pps->weighted_bipred_idc == 2 ? 0x5 | 0x5 << 3 : 0),
+		| (pps->weighted_bipred_idc == 2 ? DEFAULT_WEIGHT_DENOM : 0),
 		"slc_76c_cmd_weights_denom");
 
 	if (!pred_weight)
 		return;
 
-	default_luma_weight = 1 << weights->luma_log2_weight_denom;
-	default_chroma_weight = 1 << weights->chroma_log2_weight_denom;
+	luma_denom = 1 << weights->luma_log2_weight_denom;
+	chroma_denom = 1 << weights->chroma_log2_weight_denom;
 
 
 	for (int y = 0; y < 2; y++) {
@@ -307,40 +321,40 @@ static void stream_weights(struct avd_ctx *ctx, struct avd_h264_run *run)
 		for (int i = 0; i < to + 1; i++) {
 			/* Avd only expects offsets/weights if they are not the default
 			 * ones, otherwise we get artifacts */
-			if((factors.luma_weight[i] != default_luma_weight)
+			if((factors.luma_weight[i] != luma_denom)
 					|| (factors.luma_offset[i] != 0)) {
-				push(0x2de00000
-						| 1 << 14
-						| y << 13
-						| i << 9
-						| ((factors.luma_weight[i] & 0x1ff)),
+				push(AVD_OP_WEIGHTS
+						| AVD_OP_WEIGHTS_IDENT(1)
+						| AVD_OP_WEIGHTS_LIST_IDX(y)
+						| AVD_OP_WEIGHTS_INDEX(i)
+						| AVD_OP_WEIGHTS_WEIGHT(factors.luma_weight[i]),
 					"slc_luma_weights");
-				push(0x2df00000
-						| swrap(factors.luma_offset[i], 0x10000),
+				push(AVD_OP_OFFSETS
+						| AVD_OP_OFFSETS_OFFSET(factors.luma_offset[i]),
 						"slc_luma_offsets");
 			}
 
-			if ((factors.chroma_weight[i][0] != default_chroma_weight)
+			if ((factors.chroma_weight[i][0] != chroma_denom)
 					|| (factors.chroma_offset[i][0] != 0)
-					|| (factors.chroma_weight[i][1] != default_chroma_weight)
+					|| (factors.chroma_weight[i][1] != chroma_denom)
 					|| (factors.chroma_offset[i][1] != 0)) {
-				push(0x2de00000
-						| 2 << 14
-						| y << 13
-						| i << 9
-						| (factors.chroma_weight[i][0] & 0x1ff),
+				push(AVD_OP_WEIGHTS
+						| AVD_OP_WEIGHTS_IDENT(2)
+						| AVD_OP_WEIGHTS_LIST_IDX(y)
+						| AVD_OP_WEIGHTS_INDEX(i)
+						| AVD_OP_WEIGHTS_WEIGHT(factors.chroma_weight[i][0]),
 						"slc_chroma_weights[0]");
-				push(0x2df00000
-						| swrap(factors.chroma_offset[i][0], 0x10000),
+				push(AVD_OP_OFFSETS
+						| AVD_OP_OFFSETS_OFFSET(factors.chroma_offset[i][0]),
 						"slc_chroma_offsets[0]");
-				push(0x2de00000
-						| 3 << 14
-						| y << 13
-						| i << 9
-						| (factors.chroma_weight[i][1] & 0x1ff),
+				push(AVD_OP_WEIGHTS
+						| AVD_OP_WEIGHTS_IDENT(3)
+						| AVD_OP_WEIGHTS_LIST_IDX(y)
+						| AVD_OP_WEIGHTS_INDEX(i)
+						| AVD_OP_WEIGHTS_WEIGHT(factors.chroma_weight[i][1]),
 						"slc_chroma_weights[1]");
-				push(0x2df00000
-						| swrap(factors.chroma_offset[i][1], 0x10000),
+				push(AVD_OP_OFFSETS
+						| AVD_OP_OFFSETS_OFFSET(factors.chroma_offset[i][1]),
 						"slc_chroma_offsets[1]");
 			}
 		}
@@ -359,7 +373,6 @@ static u32 stream_slice(struct avd_ctx *ctx, struct avd_h264_run *run)
 	bool en_mode = (pps->flags & V4L2_H264_PPS_FLAG_ENTROPY_CODING_MODE) ==
 		       0;
 	const u8 *data = vb2_plane_vaddr(&src->vb2_buf, 0);
-	u32 x;
 
 	u32 min_off = (sl->header_bit_size + (en_mode ? 0 : 7)) / 8;
 
@@ -378,32 +391,29 @@ static u32 stream_slice(struct avd_ctx *ctx, struct avd_h264_run *run)
 
 	dma_addr_t slc_a84 = run->addresses.sl + off;
 
-	push(0x2d800000
-			| ((en_mode ? (sl->header_bit_size % 8) : 0) << 15)
-			| (u32)(slc_a84 >> 32), "slc_a7c_cmd_set_coded_slice");
+	push(AVD_OP_CODED_DATA
+			| AVD_OP_CODED_DATA_BIT_OFF(en_mode ?
+				(sl->header_bit_size % 8) : 0)
+			| AVD_OP_CODED_DATA_ADDR(slc_a84 >> 32), "slc_a7c_cmd_set_coded_slice");
 	push((u32)(slc_a84 & 0xffffffff), "slc_a84_slice_addr_low");
 
 	/* should not include trailing 0? */
 	push(payload_len - off, "slc_a88_slice_hdr_size");
 
-	push(0x2c000000
-			| (sl->first_mb_in_slice / (sps->pic_width_in_mbs_minus1 + 1) << 12)
-			|  sl->first_mb_in_slice % (sps->pic_width_in_mbs_minus1 + 1),
+	push(AVD_OP_SL_LOC
+			| AVD_OP_SL_LOC_Y(sl->first_mb_in_slice / (sps->pic_width_in_mbs_minus1 + 1))
+			| AVD_OP_SL_LOC_X(sl->first_mb_in_slice % (sps->pic_width_in_mbs_minus1 + 1)),
 			"cm3_cmd_exec_mb_vp");
 
-	push(0x2d900000
-		| ( ((26 + pps->pic_init_qp_minus26 + sl->slice_qp_delta) * 0x400)
-				& 0x1fc00),
+	push(AVD_OP_QP
+		| AVD_OP_QP_VAL(26 + pps->pic_init_qp_minus26 + sl->slice_qp_delta),
 			"slc_a70_cmd_quant_param");
 
-	push(0x2da00000
-			| (u32)(sl->disable_deblocking_filter_idc == 0
-				? BIT(17) : 0)
-			| (u32)(sl->disable_deblocking_filter_idc != 1
-				? BIT(16)
-				| swrap(sl->slice_beta_offset_div2, 16) << 12
-				| swrap(sl->slice_alpha_c0_offset_div2, 16) << 8
-				: 0),
+	push(AVD_OP_DBLK
+			| AVD_OP_DBLK_FLAG_FULL_EN(sl->disable_deblocking_filter_idc == 0)
+			| AVD_OP_DBLK_FLAG_EN(sl->disable_deblocking_filter_idc != 1)
+			| AVD_OP_DBLK_OFF1(sl->slice_beta_offset_div2)
+			| AVD_OP_DBLK_OFF0(sl->slice_alpha_c0_offset_div2),
 			"slc_a74_cmd_deblocking_filter");
 
 	if (sl->slice_type == V4L2_H264_SLICE_TYPE_P
@@ -411,52 +421,42 @@ static u32 stream_slice(struct avd_ctx *ctx, struct avd_h264_run *run)
 
 		u32 num_ref_idx_active = sl->num_ref_idx_l0_active_minus1 + 1;
 		for (u32 i = 0; i < num_ref_idx_active; i++)
-			push(0x2dc00000
-			   | (0 << 8)
-			   | ((i & 0xf) << 4)
-			   | (sl->ref_pic_list0[i].index & 0xf),
+			push(AVD_OP_REF
+			   | AVD_OP_REF_LIST_IDX(0)
+			   | AVD_OP_REF_LOOP_IDX(i)
+			   | AVD_OP_REF_DBP_IDX(sl->ref_pic_list0[i].index),
 			   "slc_6e8_cmd_ref_list_0");
 
 		if (sl->slice_type == V4L2_H264_SLICE_TYPE_B) {
 			u32 num_ref_idx_active = sl->num_ref_idx_l1_active_minus1 + 1;
 			for (u32 i = 0; i < num_ref_idx_active; i++)
-				push(0x2dc00000
-						| (1 << 8)
-						| ((i & 0xf) << 4)
-						| (sl->ref_pic_list1[i].index & 0xf),
+				push(AVD_OP_REF
+						| AVD_OP_REF_LIST_IDX(1)
+						| AVD_OP_REF_LOOP_IDX(i)
+						| AVD_OP_REF_DBP_IDX(sl->ref_pic_list1[i].index),
 						"slc_6e8_cmd_ref_list_0");
 		}
 		stream_weights(ctx, run);
 	}
 
 	if (sl->first_mb_in_slice == 0) {
-		push(0x2a000000, "cm3_cmd_set_mb_dims");
-		push(((sps->pic_height_in_map_units_minus1) << 12)
-				| (sps->pic_width_in_mbs_minus1), "cm3_set_mb_dims");
+		push(AVD_OP_SL_DIM_START, "cm3_cmd_set_mb_dims");
+		push(AVD_SL_DIM_END_Y(sps->pic_height_in_map_units_minus1)
+				| AVD_SL_DIM_END_X(sps->pic_width_in_mbs_minus1), "cm3_set_mb_dims");
 	}
 
-	x = 0;
-	if (sl->slice_type == V4L2_H264_SLICE_TYPE_I)
-		x |= 0x20000;
-	else if (sl->slice_type == V4L2_H264_SLICE_TYPE_P)
-		x |= 0x10000;
-	else if (sl->slice_type == V4L2_H264_SLICE_TYPE_B)
-		x |= 0x40000;
+	push(AVD_OP_SL_REF |
+			AVD_OP_SL_REF_FLAG_CABAC(sl->cabac_init_idc == 1) |
+			AVD_OP_SL_REF_FLAG1(sl->cabac_init_idc == 2) |
+			AVD_OP_SL_REF_FLAG2(!(sl->flags &
+			      V4L2_H264_SLICE_FLAG_DIRECT_SPATIAL_MV_PRED)) |
+			AVD_OP_SL_REF_NUM_L0(sl->num_ref_idx_l0_active_minus1) |
+			AVD_OP_SL_REF_NUM_L1(sl->num_ref_idx_l1_active_minus1) |
+			AVD_OP_SL_REF_SLICE_P(sl->slice_type == V4L2_H264_SLICE_TYPE_P) |
+			AVD_OP_SL_REF_SLICE_I(sl->slice_type == V4L2_H264_SLICE_TYPE_I) |
+			AVD_OP_SL_REF_SLICE_B(sl->slice_type == V4L2_H264_SLICE_TYPE_B)
 
-	if ((sl->slice_type == V4L2_H264_SLICE_TYPE_P) ||
-	    (sl->slice_type == V4L2_H264_SLICE_TYPE_B)) {
-		if (pps->flags & V4L2_H264_PPS_FLAG_ENTROPY_CODING_MODE)
-			x |= sl->cabac_init_idc << 5;
-
-		if (sl->slice_type == V4L2_H264_SLICE_TYPE_B) {
-			x |= sl->num_ref_idx_l1_active_minus1 << 7;
-			if (!(sl->flags &
-			      V4L2_H264_SLICE_FLAG_DIRECT_SPATIAL_MV_PRED))
-				x |= BIT(15);
-		}
-		x |= sl->num_ref_idx_l0_active_minus1 << 11;
-	}
-	push(0x2d000000 | x, "slc_6e4_cmd_ref_type");
+			, "slc_6e4_cmd_ref_type");
 
 	if (sl->slice_type == V4L2_H264_SLICE_TYPE_B) {
 		/* bidirectional reference of previous mv */
@@ -473,8 +473,8 @@ static u32 stream_slice(struct avd_ctx *ctx, struct avd_h264_run *run)
 	}
 
 	/* only submit if this is the last slice */
-	push(0x2b000000
-			| !(src->flags & V4L2_BUF_FLAG_M2M_HOLD_CAPTURE_BUF) << 10,
+	push(AVD_OP_EXEC
+			| AVD_OP_EXEC_FLAG_END(!(src->flags & V4L2_BUF_FLAG_M2M_HOLD_CAPTURE_BUF)),
 			"cm3_cmd_inst_fifo_end");
 
 	return payload_len - off;
diff --git a/drivers/media/platform/apple/avd/avd-hevc.c b/drivers/media/platform/apple/avd/avd-hevc.c
index 478db2d9eca076..57172067a17657 100644
--- a/drivers/media/platform/apple/avd/avd-hevc.c
+++ b/drivers/media/platform/apple/avd/avd-hevc.c
@@ -13,6 +13,7 @@
  *	Tomasz Figa <tfiga@chromium.org>
  */
 
+#include "asm-generic/errno-base.h"
 #include "linux/dev_printk.h"
 #include <linux/v4l2-controls.h>
 #include <linux/delay.h>
@@ -28,6 +29,29 @@
 #define NEW_TILE_ID BIT(0)
 #define NEW_SLICE BIT(1)
 
+#define HEVC_PCM_EN(v)		FIELD_PREP(BIT(12), !!(v))
+#define HEVC_PCM_BD_LUMA(v)	FIELD_PREP(GENMASK(11, 8), v)
+#define HEVC_PCM_BD_CHROMA(v)	FIELD_PREP(GENMASK(7, 4), v)
+#define HEVC_PCM_CB_MIN_LUMA(v)	FIELD_PREP(GENMASK(3, 2), v)
+#define HEVC_PCM_CB_LUMA(v)	FIELD_PREP(GENMASK(1, 0), v)
+
+#define HEVC_UNK_ISM_EN(v)	FIELD_PREP(BIT(9), !!(v))
+#define HEVC_UNK_FLAG		FIELD_PREP(BIT(3), 1)
+
+#define HEVC_CTB_SIZE(v)	FIELD_PREP(GENMASK(8, 3), v)
+#define HEVC_MERGE_LV(v)	FIELD_PREP(GENMASK(11, 9), v)
+#define HEVC_FLAG_ENTROPY_EN(v)	FIELD_PREP(BIT(12), !!(v))
+#define HEVC_FLAG_TILES_EN(v)	FIELD_PREP(BIT(13), !!(v))
+#define HEVC_FLAG_TQB_EN(v)	FIELD_PREP(BIT(14), !!(v))
+#define HEVC_CU_QP_DD(v)	FIELD_PREP(GENMASK(16, 15), v)
+#define HEVC_FLAG_CU_QP_EN(v)	FIELD_PREP(BIT(17), !!(v))
+#define HEVC_FLAG_TSKIP_EN(v)	FIELD_PREP(BIT(18), !!(v))
+#define HEVC_FLAG_CI_PRED(v)	FIELD_PREP(BIT(19), !!(v))
+#define HEVC_FLAG_SDH_EN(v)	FIELD_PREP(BIT(20), !!(v))
+#define HEVC_FLAG_TMVP_EN(v)	FIELD_PREP(BIT(21), !!(v))
+
+#define HEVC_SCL_DIMS 0x127ffff
+
 static inline u32 sps_size(u32 w, u32 h)
 {
 	/* TODO: does it really need so much? */
@@ -43,6 +67,7 @@ struct avd_hevc_run {
 	const struct v4l2_ctrl_hevc_scaling_matrix *scaling_matrix;
 	const u32 *entry_point_offsets;
 
+	int num_entry_point_offsets;
 	int num_slices;
 
 	struct run_addr {
@@ -97,10 +122,10 @@ static void stream_refs(struct avd_ctx *ctx, struct avd_hevc_run *run)
 			vb2_dma_contig_plane_dma_addr(&ref_buf->base.vb.vb2_buf, 0)
 				+ (ref_buf->base.vb.planes[0].length - sps_size(fmt_width(ctx), fmt_height(ctx)) - ref_buf->rvra.size);
 
-		push((decode->num_active_dpb_entries - 1) << 28
-				| 0x1000000
-				| (!!(dpb->flags & V4L2_HEVC_DPB_ENTRY_LONG_TERM_REFERENCE) << 17)
-				| (((sl->slice_pic_order_cnt - dpb->pic_order_cnt_val) & 0x1ffff)),
+		push(AVD_REF_NUM(decode->num_active_dpb_entries - 1)
+				| AVD_REF_FLAG_CONST
+				| AVD_REF_FLAG_LONG(dpb->flags & V4L2_HEVC_DPB_ENTRY_LONG_TERM_REFERENCE)
+				| AVD_REF_DELTA_POC(sl->slice_pic_order_cnt - dpb->pic_order_cnt_val),
 				"hdr_d0_ref_hdr");
 
 		push_rvra(avd, ctx, rvra_addr, ref_buf->rvra.offsets);
@@ -123,48 +148,48 @@ static void set_scaling_lists(struct avd_ctx *ctx, struct avd_hevc_run *run)
 	 * this should presumably be how many of each are enabled?
 	 * or some other scaling related thing
 	 */
-	push(0x127ffff, "hdr_7c_pps_scl_dims");
+	push(HEVC_SCL_DIMS, "hdr_7c_pps_scl_dims");
 
 	for (i = 0; i < 2; i++)
-		push(dc_16x16[i][0] << 16
-				| dc_16x16[i][1] << 8
-				| dc_16x16[i][2] << 0, "dc_16x16");
+		push(AVD_SCALING_I2(dc_16x16[i][0])
+				| AVD_SCALING_I1(dc_16x16[i][1])
+				| AVD_SCALING_I0(dc_16x16[i][2]), "dc_16x16");
 
 	for (i = 0; i < 2; i++)
-		push(s->scaling_list_dc_coef_32x32[i] << 16, "dc_32x32");
+		push(AVD_SCALING_I2(s->scaling_list_dc_coef_32x32[i]), "dc_32x32");
 
 	/* transposed in stride 4 */
 	for (i = 0; i < 6; i++)
 		for (j = 0; j < 4; j++)
-			push(sc_4x4[i][0][j] << 24
-					| sc_4x4[i][1][j] << 16
-					| sc_4x4[i][2][j] << 8
-					| sc_4x4[i][3][j], "scaling_4x4");
+			push(AVD_SCALING_I3(sc_4x4[i][0][j])
+					| AVD_SCALING_I2(sc_4x4[i][1][j])
+					| AVD_SCALING_I1(sc_4x4[i][2][j])
+					| AVD_SCALING_I0(sc_4x4[i][3][j]), "scaling_4x4");
 
 	/* transposed in stride 8 */
 	for (i = 0; i < 6; i++)
 		for (j = 0; j < 2; j++)
 			for (k = 0; k < 8; k++)
-			push(sc_8x8[i][j][0][k] << 24
-					| sc_8x8[i][j][1][k] << 16
-					| sc_8x8[i][j][2][k] << 8
-					| sc_8x8[i][j][3][k], "scaling_8x8");
+			push(AVD_SCALING_I3(sc_8x8[i][j][0][k])
+					| AVD_SCALING_I2(sc_8x8[i][j][1][k])
+					| AVD_SCALING_I1(sc_8x8[i][j][2][k])
+					| AVD_SCALING_I0(sc_8x8[i][j][3][k]), "scaling_8x8");
 
 	for (i = 0; i < 6; i++)
 		for (j = 0; j < 2; j++)
 			for (k = 0; k < 8; k++)
-			push(sc_16x16[i][j][0][k] << 24
-					| sc_16x16[i][j][1][k] << 16
-					| sc_16x16[i][j][2][k] << 8
-					| sc_16x16[i][j][3][k], "scaling_16x16");
+			push(AVD_SCALING_I3(sc_16x16[i][j][0][k])
+					| AVD_SCALING_I2(sc_16x16[i][j][1][k])
+					| AVD_SCALING_I1(sc_16x16[i][j][2][k])
+					| AVD_SCALING_I0(sc_16x16[i][j][3][k]), "scaling_16x16");
 
 	for (i = 0; i < 2; i++)
 		for (j = 0; j < 2; j++)
 			for (k = 0; k < 8; k++)
-				push(sc_32x32[i][j][0][k] << 24
-						| sc_32x32[i][j][1][k] << 16
-						| sc_32x32[i][j][2][k] << 8
-						| sc_32x32[i][j][3][k], "scaling_32x32");
+				push(AVD_SCALING_I3(sc_32x32[i][j][0][k])
+						| AVD_SCALING_I2(sc_32x32[i][j][1][k])
+						| AVD_SCALING_I1(sc_32x32[i][j][2][k])
+						| AVD_SCALING_I0(sc_32x32[i][j][3][k]), "scaling_32x32");
 
 }
 
@@ -175,43 +200,40 @@ static void hevc_set_flags(struct avd_ctx *ctx, struct avd_hevc_run *run)
 	const struct v4l2_ctrl_hevc_pps *pps = run->pps;
 	struct avd_dev *avd = ctx->dev;
 
-	u32 log2_ctb_size = ((sps->log2_min_luma_coding_block_size_minus3 + 3) +
+	u32 log2_ctb_size = ((sps->log2_min_luma_coding_block_size_minus3) +
 				sps->log2_diff_max_min_luma_coding_block_size);
 
-	if (sps->flags & V4L2_HEVC_SPS_FLAG_PCM_ENABLED)
-		push(0x1000
-			| sps->pcm_sample_bit_depth_luma_minus1 << 8
-			| sps->pcm_sample_bit_depth_chroma_minus1 << 4
-			| sps->log2_min_pcm_luma_coding_block_size_minus3 << 2
-			| (sps->log2_diff_max_min_pcm_luma_coding_block_size
-			  + sps->log2_min_pcm_luma_coding_block_size_minus3)
+	push(HEVC_PCM_EN(sps->flags & V4L2_HEVC_SPS_FLAG_PCM_ENABLED)
+			| HEVC_PCM_BD_LUMA(sps->pcm_sample_bit_depth_luma_minus1)
+			| HEVC_PCM_BD_CHROMA(sps->pcm_sample_bit_depth_chroma_minus1)
+			| HEVC_PCM_CB_MIN_LUMA(sps->log2_min_pcm_luma_coding_block_size_minus3)
+			| HEVC_PCM_CB_LUMA(sps->log2_diff_max_min_pcm_luma_coding_block_size
+				+ sps->log2_min_pcm_luma_coding_block_size_minus3)
 			, "hdr_30_sps_pcm");
-	else
-		push(0, "hdr_30_sps_pcm");
 
 	/*
 	 * RExt sets a few new flags here
 	 */
-	push((u32)BIT(3)
-		| !!(sps->flags & V4L2_HEVC_SPS_FLAG_STRONG_INTRA_SMOOTHING_ENABLED) << 9,
+	push(HEVC_UNK_FLAG
+		| HEVC_UNK_ISM_EN(sps->flags & V4L2_HEVC_SPS_FLAG_STRONG_INTRA_SMOOTHING_ENABLED),
 			"hdr_34_sps_flags");
 
-	push((log2_ctb_size - 3) << 3
-		| pps->log2_parallel_merge_level_minus2 << 9
-		| !!(pps->flags & V4L2_HEVC_PPS_FLAG_ENTROPY_CODING_SYNC_ENABLED) << 12
-		| !!(pps->flags & V4L2_HEVC_PPS_FLAG_TILES_ENABLED) << 13
-		| !!(pps->flags & V4L2_HEVC_PPS_FLAG_TRANSQUANT_BYPASS_ENABLED) << 14
-		| ((log2_ctb_size - pps->diff_cu_qp_delta_depth - 3) & 3) << 15
-		| !!(pps->flags & V4L2_HEVC_PPS_FLAG_CU_QP_DELTA_ENABLED) << 17
-		| !!(pps->flags & V4L2_HEVC_PPS_FLAG_TRANSFORM_SKIP_ENABLED) << 18
-		| !!(pps->flags & V4L2_HEVC_PPS_FLAG_CONSTRAINED_INTRA_PRED) << 19
-		| !!(pps->flags & V4L2_HEVC_PPS_FLAG_SIGN_DATA_HIDING_ENABLED) << 20
-		| !!(!(decode->flags & V4L2_HEVC_DECODE_PARAM_FLAG_IDR_PIC)
-			&& sps->flags & V4L2_HEVC_SPS_FLAG_SPS_TEMPORAL_MVP_ENABLED) << 21
+	push(HEVC_CTB_SIZE(log2_ctb_size)
+		| HEVC_MERGE_LV(pps->log2_parallel_merge_level_minus2)
+		| HEVC_FLAG_ENTROPY_EN(pps->flags & V4L2_HEVC_PPS_FLAG_ENTROPY_CODING_SYNC_ENABLED)
+		| HEVC_FLAG_TILES_EN(pps->flags & V4L2_HEVC_PPS_FLAG_TILES_ENABLED)
+		| HEVC_FLAG_TQB_EN(pps->flags & V4L2_HEVC_PPS_FLAG_TRANSQUANT_BYPASS_ENABLED)
+		| HEVC_CU_QP_DD(log2_ctb_size - pps->diff_cu_qp_delta_depth)
+		| HEVC_FLAG_CU_QP_EN(pps->flags & V4L2_HEVC_PPS_FLAG_CU_QP_DELTA_ENABLED)
+		| HEVC_FLAG_TSKIP_EN(pps->flags & V4L2_HEVC_PPS_FLAG_TRANSFORM_SKIP_ENABLED)
+		| HEVC_FLAG_CI_PRED(pps->flags & V4L2_HEVC_PPS_FLAG_CONSTRAINED_INTRA_PRED)
+		| HEVC_FLAG_SDH_EN(pps->flags & V4L2_HEVC_PPS_FLAG_SIGN_DATA_HIDING_ENABLED)
+		| HEVC_FLAG_TMVP_EN(!(decode->flags & V4L2_HEVC_DECODE_PARAM_FLAG_IDR_PIC)
+			&& sps->flags & V4L2_HEVC_SPS_FLAG_SPS_TEMPORAL_MVP_ENABLED)
 		, "hdr_5c_pps_flags");
 
-	push(swrap(pps->pps_cb_qp_offset, 1 << 5) << 5
-			| swrap(pps->pps_cr_qp_offset, 1 << 5),
+	push(AVD_HDR_H26X_QP_OFFSET_CB(pps->pps_cb_qp_offset)
+			| AVD_HDR_H26X_QP_OFFSET_CR(pps->pps_cr_qp_offset),
 			"hdr_60_pps_qp");
 
 	/*
@@ -237,42 +259,47 @@ static void set_header(struct avd_ctx *ctx, struct avd_hevc_run *run)
 
 	bool is_intra = run->sl[0].slice_type == V4L2_HEVC_SLICE_TYPE_I;
 
-	push(0x2b000000
-		| ctx->fifo_idx << 4
-		| (avd->variant->revision == 3 ? 0x100 : 0x200),
+	push(AVD_OP_EXEC
+		| AVD_OP_EXEC_FIFO_IDX(ctx->fifo_idx)
+		| AVD_OP_EXEC_FLAG_START_REV3(avd->variant->revision == 3)
+		| AVD_OP_EXEC_FLAG_START_REV4(avd->variant->revision == 4),
 		"inst_fifo_start");
 
-	push(0x2db00000
-		| 0x1000
-		| (is_intra ? 0x2000 : 0)
-		| 0x2e0
-		| (avd->variant->quirks & AVD_QUIRK_NO_PIPE_STATE ? 0 : 0x80000),
+	push(AVD_OP_HDR
+		| AVD_OP_HDR_FLAG0
+		| AVD_OP_HDR_FLAG_INTRA(is_intra)
+		| AVD_OP_HDR_CONST
+		| AVD_OP_HDR_FLAG_PIPE_STATE(!(avd->variant->quirks & AVD_QUIRK_NO_PIPE_STATE)),
 		"hdr_34_start_hdr");
 
-	push(AVD_CODEC_HEVC, "hdr_50_mode");
-	push(((height - 1) << 16) | (width - 1), "hdr_54_height_width");
+	push(AVD_HDR_CODEC_MODE(AVD_CODEC_HEVC), "hdr_50_mode");
+	push(AVD_HDR_HEIGHT(height - 1) | AVD_HDR_WIDTH(width - 1),
+			"hdr_54_height_width");
 	push(0, "hdr_58_pixfmt_zero");
 
-	push((((height - 1) >> 3) << 16) | ((width - 1) >> 3),
+	push(AVD_HDR_HEIGHT((height - 1) >> 3) |
+			AVD_HDR_WIDTH((width - 1) >> 3),
 			"hdr_28_height_width_shift3");
 
-	push(sps->chroma_format_idc << 24
-		| sps->bit_depth_chroma_minus8 << 19
-		| sps->bit_depth_luma_minus8 << 15
-		| sps->log2_min_luma_coding_block_size_minus3 << 13
-		| (sps->log2_diff_max_min_luma_coding_block_size +
-			sps->log2_min_luma_coding_block_size_minus3) << 11
-		| sps->log2_min_luma_transform_block_size_minus2 << 9
-		| (sps->log2_diff_max_min_luma_transform_block_size +
-			+ sps->log2_min_luma_transform_block_size_minus2) << 7
-		| sps->max_transform_hierarchy_depth_inter << 4
-		| sps->max_transform_hierarchy_depth_intra << 1
-		| !!(sps->flags & V4L2_HEVC_SPS_FLAG_AMP_ENABLED),
+	push(AVD_HDR_COMMON_CHROMA_FORMAT(sps->chroma_format_idc)
+		| AVD_HDR_COMMON_BIT_DEPTH_C(sps->bit_depth_chroma_minus8)
+		| AVD_HDR_COMMON_BIT_DEPTH_L(sps->bit_depth_luma_minus8)
+		| AVD_HDR_COMMON_MIN_LUMA_CBS(sps->log2_min_luma_coding_block_size_minus3)
+		| AVD_HDR_COMMON_LUMA_CBS(sps->log2_diff_max_min_luma_coding_block_size +
+			sps->log2_min_luma_coding_block_size_minus3)
+		| AVD_HDR_COMMON_MIN_LUMA_TBS(sps->log2_min_luma_transform_block_size_minus2)
+		| AVD_HDR_COMMON_LUMA_TBS(sps->log2_diff_max_min_luma_transform_block_size +
+			+ sps->log2_min_luma_transform_block_size_minus2)
+		| AVD_HDR_COMMON_TR_INTER(sps->max_transform_hierarchy_depth_inter)
+		| AVD_HDR_COMMON_TR_INTRA(sps->max_transform_hierarchy_depth_intra)
+		| AVD_HDR_COMMON_FLAG0(sps->flags & V4L2_HEVC_SPS_FLAG_AMP_ENABLED),
 		 "hdr_2c_sps_txfm");
 
 	hevc_set_flags(ctx, run);
-	push(0x300000
-			| (avd->variant->quirks & AVD_QUIRK_NO_PIPE_STATE ? 0 : 0x30),
+
+	push(AVD_HDR_FEAT_H26X | AVD_HDR_FEAT_COMMON |
+			AVD_HDR_FEAT_PIPE_STATE_EN(!(avd->variant->quirks &
+					AVD_QUIRK_NO_PIPE_STATE)),
 			"hdr_98_const_30");
 
 	push(INST_DMA2, "cm3_dma_config_1");
@@ -323,7 +350,8 @@ static void set_header(struct avd_ctx *ctx, struct avd_hevc_run *run)
 	pusha(run->addresses.uv, "hdr_1b8_uv_addr_lsb8", 0);
 	push(bytesperline, "hdr_1c0_width_align");
 	push(0, "");
-	push(((height - 1) << 16) | (width - 1), "hdr_54_height_width");
+	push(AVD_HDR_HEIGHT(height - 1) | AVD_HDR_WIDTH(width - 1),
+			"hdr_54_height_width");
 
 	if (!is_intra)
 		stream_refs(ctx, run);
@@ -354,22 +382,18 @@ static void stream_weights(struct avd_ctx *ctx, struct avd_hevc_run *run,
 	const s8 *luma_offsets;
 
 	if (!has_luma_weights) {
-		push(0x2dd00000, "slc_76c_cmd_weights_denom");
+		push(AVD_OP_WEIGHTS_HDR, "slc_76c_cmd_weights_denom");
 		return;
 	}
 
 	chroma_log2_weight_denom = pred->luma_log2_weight_denom +
 		pred->delta_chroma_log2_weight_denom;
 
-	/*
-	 * TODO: this looks weird
-	 */
-	push(0x2dd00000
-			| (!has_luma_weights << 7)
-			| (has_luma_weights << 6)
-			| (!has_luma_weights << 5)
-			| (pred->luma_log2_weight_denom << 3)
-			| (chroma_log2_weight_denom & 7)
+	push(AVD_OP_WEIGHTS_HDR |
+			AVD_OP_WEIGHTS_HDR_FLAG1(!has_luma_weights) |
+			AVD_OP_WEIGHTS_HDR_FLAG0(has_luma_weights) |
+			AVD_OP_WEIGHTS_HDR_LUMA(pred->luma_log2_weight_denom) |
+			AVD_OP_WEIGHTS_HDR_CHROMA(chroma_log2_weight_denom)
 			,"slc_76c_cmd_weights_denom");
 
 	luma_weight_denom = 1 << pred->luma_log2_weight_denom;
@@ -394,14 +418,14 @@ static void stream_weights(struct avd_ctx *ctx, struct avd_hevc_run *run,
 			/* Avd only expects offsets/weights if they are not the default
 			 * ones, otherwise we get artifacts */
 			if((delta_luma_weights[i]  != 0) || (luma_offsets[i] != 0)) {
-				push(0x2de00000
-						| 1 << 14
-						| y << 13
-						| i << 9
-						| ((delta_luma_weights[i] + luma_weight_denom) & 0x1ff),
+				push(AVD_OP_WEIGHTS
+						| AVD_OP_WEIGHTS_IDENT(1)
+						| AVD_OP_WEIGHTS_LIST_IDX(y)
+						| AVD_OP_WEIGHTS_INDEX(i)
+						| AVD_OP_WEIGHTS_WEIGHT(delta_luma_weights[i] + luma_weight_denom),
 						"slc_luma_weights");
-				push(0x2df00000
-						| swrap(luma_offsets[i], 0x10000),
+				push(AVD_OP_OFFSETS
+						| AVD_OP_OFFSETS_OFFSET(luma_offsets[i]),
 						"slc_luma_offsets");
 			}
 
@@ -409,23 +433,23 @@ static void stream_weights(struct avd_ctx *ctx, struct avd_hevc_run *run,
 					|| (chroma_offsets[i][0] != 0)
 					|| (delta_chroma_weights[i][1] != 0)
 					|| (chroma_offsets[i][1] != 0)) {
-				push(0x2de00000
-						| 2 << 14
-						| y << 13
-						| i << 9
-						| ((delta_chroma_weights[i][0] + chroma_weight_denom) & 0x1ff),
+				push(AVD_OP_WEIGHTS
+						| AVD_OP_WEIGHTS_IDENT(2)
+						| AVD_OP_WEIGHTS_LIST_IDX(y)
+						| AVD_OP_WEIGHTS_INDEX(i)
+						| AVD_OP_WEIGHTS_WEIGHT(delta_chroma_weights[i][0] + chroma_weight_denom),
 						"slc_chroma_weights[0]");
-				push(0x2df00000
-						| swrap(chroma_offsets[i][0], 0x10000),
+				push(AVD_OP_OFFSETS
+						| AVD_OP_OFFSETS_OFFSET(chroma_offsets[i][0]),
 						"slc_chroma_offsets[0]");
-				push(0x2de00000
-						| 3 << 14
-						| y << 13
-						| i << 9
-						| ((delta_chroma_weights[i][1] + chroma_weight_denom) & 0x1ff),
+				push(AVD_OP_WEIGHTS
+						| AVD_OP_WEIGHTS_IDENT(3)
+						| AVD_OP_WEIGHTS_LIST_IDX(y)
+						| AVD_OP_WEIGHTS_INDEX(i)
+						| AVD_OP_WEIGHTS_WEIGHT(delta_chroma_weights[i][1] + chroma_weight_denom),
 						"slc_chroma_weights[1]");
-				push(0x2df00000
-						| swrap(chroma_offsets[i][1], 0x10000),
+				push(AVD_OP_OFFSETS
+						| AVD_OP_OFFSETS_OFFSET(chroma_offsets[i][1]),
 						"slc_chroma_offsets[1]");
 			}
 		}
@@ -439,43 +463,43 @@ static void stream_slice_dqtblk(struct avd_ctx *ctx, struct avd_hevc_run *run,
 	const struct v4l2_ctrl_hevc_sps *sps = run->sps;
 	struct avd_dev *avd = ctx->dev;
 
-	push(0x2d900000
-			| (((pps->init_qp_minus26 + 26 +  sl->slice_qp_delta) << 10) & 0x1fc00)
-			| swrap(pps->pps_cb_qp_offset + sl->slice_cb_qp_offset, 32) << 5
-			| swrap(pps->pps_cr_qp_offset + sl->slice_cr_qp_offset, 32)
+	push(AVD_OP_QP |
+			AVD_OP_QP_VAL(pps->init_qp_minus26 + 26 +  sl->slice_qp_delta) |
+			AVD_OP_QP_CB_OFF(pps->pps_cb_qp_offset + sl->slice_cb_qp_offset) |
+			AVD_OP_QP_CR_OFF(pps->pps_cr_qp_offset + sl->slice_cr_qp_offset)
 			,"slc_bcc_cmd_quantization");
 
-	push(0x2da00000
-			| !!(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_SLICE_SAO_CHROMA) << 6
-		    | !!(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_SLICE_SAO_LUMA) << 7
-			| swrap(sl->slice_tc_offset_div2, 16) << 8
-			| swrap(sl->slice_beta_offset_div2, 16) << 12
+	push(AVD_OP_DBLK |
+			AVD_OP_DBLK_FLAG_SAO_CHROMA(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_SLICE_SAO_CHROMA) |
+			AVD_OP_DBLK_FLAG_SAO_LUMA(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_SLICE_SAO_LUMA) |
+			AVD_OP_DBLK_OFF0(sl->slice_tc_offset_div2) |
+			AVD_OP_DBLK_OFF1(sl->slice_beta_offset_div2) |
 			/* i wonder what this should actually be */
-			| !!(!(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_SLICE_DEBLOCKING_FILTER_DISABLED)
+			AVD_OP_DBLK_FLAG_EN(!(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_SLICE_DEBLOCKING_FILTER_DISABLED)
 				&& (sps->flags & V4L2_HEVC_SPS_FLAG_STRONG_INTRA_SMOOTHING_ENABLED
 				|| sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_SLICE_SAO_LUMA
-				|| pps->flags & (V4L2_HEVC_PPS_FLAG_DEBLOCKING_FILTER_OVERRIDE_ENABLED | V4L2_HEVC_PPS_FLAG_DEBLOCKING_FILTER_CONTROL_PRESENT))) << 16
-			| !!(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_SLICE_LOOP_FILTER_ACROSS_SLICES_ENABLED) << 17
-			| !!(!(pps->flags & V4L2_HEVC_PPS_FLAG_TILES_ENABLED)
-				|| (pps->flags & V4L2_HEVC_PPS_FLAG_LOOP_FILTER_ACROSS_TILES_ENABLED)) << 18
-			| !!((sps->flags & V4L2_HEVC_SPS_FLAG_PCM_ENABLED)
-				&& ! (sps->flags & V4L2_HEVC_SPS_FLAG_PCM_LOOP_FILTER_DISABLED)) << 19
+				|| pps->flags & (V4L2_HEVC_PPS_FLAG_DEBLOCKING_FILTER_OVERRIDE_ENABLED | V4L2_HEVC_PPS_FLAG_DEBLOCKING_FILTER_CONTROL_PRESENT))) |
+			AVD_OP_DBLK_FLAG_FULL_EN(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_SLICE_LOOP_FILTER_ACROSS_SLICES_ENABLED) |
+			AVD_OP_DBLK_FLAG_TILES_EN(!(pps->flags & V4L2_HEVC_PPS_FLAG_TILES_ENABLED)
+				|| (pps->flags & V4L2_HEVC_PPS_FLAG_LOOP_FILTER_ACROSS_TILES_ENABLED)) |
+			AVD_OP_DBLK_FLAG_PCM_EN((sps->flags & V4L2_HEVC_SPS_FLAG_PCM_ENABLED)
+				&& ! (sps->flags & V4L2_HEVC_SPS_FLAG_PCM_LOOP_FILTER_DISABLED))
 			, "slc_bd0_cmd_deblocking_filter");
 
 	if (sl->slice_type == V4L2_HEVC_SLICE_TYPE_B
 			|| sl->slice_type == V4L2_HEVC_SLICE_TYPE_P) {
 		for (int i = 0; i < sl->num_ref_idx_l0_active_minus1 + 1; i++)
-			push(0x2dc00000
-					| (0 << 8)
-					| ((i & 0xf) << 4)
-					| (sl->ref_idx_l0[i] & 0xf),
+			push(AVD_OP_REF |
+					AVD_OP_REF_LIST_IDX(0) |
+					AVD_OP_REF_LOOP_IDX(i) |
+					AVD_OP_REF_DBP_IDX(sl->ref_idx_l0[i]),
 					"reference_frames_l0");
 		if (sl->slice_type == V4L2_HEVC_SLICE_TYPE_B)
 			for (int i = 0; i < sl->num_ref_idx_l1_active_minus1 + 1; i++)
-				push(0x2dc00000
-						| (1 << 8)
-						| ((i & 0xf) << 4)
-						| (sl->ref_idx_l1[i] & 0xf),
+				push(AVD_OP_REF |
+						AVD_OP_REF_LIST_IDX(1) |
+						AVD_OP_REF_LOOP_IDX(i) |
+						AVD_OP_REF_DBP_IDX(sl->ref_idx_l1[i]),
 						"reference_frames_l1");
 
 		stream_weights(ctx, run, sl);
@@ -491,8 +515,8 @@ static void stream_slice_mv(struct avd_ctx *ctx, struct avd_hevc_run *run,
 	const u8 *ref_list;
 
 	if (sl->slice_type == V4L2_HEVC_SLICE_TYPE_I) {
-		push(0x2d000000
-				| !!(sl->slice_type == V4L2_HEVC_SLICE_TYPE_I) << 17
+		push(AVD_OP_SL_REF |
+				AVD_OP_SL_REF_SLICE_I(sl->slice_type == V4L2_HEVC_SLICE_TYPE_I)
 				, "slc_a8c_cmd_ref_type");
 		return;
 	}
@@ -512,25 +536,18 @@ static void stream_slice_mv(struct avd_ctx *ctx, struct avd_hevc_run *run,
 		&& is_first
 		&& !ref->hevc.is_intra;
 
-	u32 x = 0;
-
-	x |= (5 - sl->five_minus_max_num_merge_cand) << 1;
-	x |= !!((sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_SLICE_TEMPORAL_MVP_ENABLED)
-			&& !(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_COLLOCATED_FROM_L0)) << 4;
-	x |= !!(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_CABAC_INIT) << 5;
-	if (sl->slice_type == V4L2_HEVC_SLICE_TYPE_B) {
-		x |= !!!(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_MVD_L1_ZERO) << 6;
-		x |= sl->num_ref_idx_l1_active_minus1 << 7;
-	}
-	x |= sl->num_ref_idx_l0_active_minus1 << 11;
-
-	x |= !!((sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_SLICE_TEMPORAL_MVP_ENABLED)
-			|| (sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_DEPENDENT_SLICE_SEGMENT)) << 15;
-
-	push(0x2d000000
-			| x
-			| !!(sl->slice_type == V4L2_HEVC_SLICE_TYPE_P) << 16
-			| ref_valid << 18
+	push(AVD_OP_SL_REF |
+			AVD_OP_SL_REF_MAX_MERGE(5 - sl->five_minus_max_num_merge_cand) |
+			AVD_OP_SL_REF_FLAG_CABAC(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_CABAC_INIT) |
+			AVD_OP_SL_REF_FLAG0((sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_SLICE_TEMPORAL_MVP_ENABLED)
+			&& !(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_COLLOCATED_FROM_L0)) |
+			AVD_OP_SL_REF_FLAG1(!(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_MVD_L1_ZERO)) |
+			AVD_OP_SL_REF_FLAG2((sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_SLICE_TEMPORAL_MVP_ENABLED)
+			|| (sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_DEPENDENT_SLICE_SEGMENT)) |
+			AVD_OP_SL_REF_NUM_L0(sl->num_ref_idx_l0_active_minus1) |
+			AVD_OP_SL_REF_NUM_L1(sl->num_ref_idx_l1_active_minus1) |
+			AVD_OP_SL_REF_SLICE_P(sl->slice_type == V4L2_HEVC_SLICE_TYPE_P) |
+			AVD_OP_SL_REF_SLICE_B(ref_valid)
 			, "slc_a8c_cmd_ref_type");
 
 	if (ref_valid) {
@@ -547,9 +564,10 @@ static void set_slice(struct avd_ctx *ctx, struct avd_hevc_run *run,
 {
 	struct avd_dev *avd = ctx->dev;
 	dma_addr_t slc_addr = run->addresses.sl + offset + sl->data_byte_offset;
-	push(0x2d800000
+	push(AVD_OP_CODED_DATA
 			| flags
-			| (u32)(slc_addr >> 32), "cm3_cmd_set_coded_slice");
+			| AVD_OP_CODED_DATA_ADDR(slc_addr >> 32),
+			"cm3_cmd_set_coded_slice");
 	push((u32)(slc_addr & 0xffffffff), "slc_bd8_slice_addr");
 	push(size, "slc_bdc_slice_size");
 }
@@ -570,7 +588,7 @@ static int submit_slice_segment(struct avd_ctx *ctx, struct avd_hevc_run *run,
 		tb_x = sl->slice_segment_addr % pic_in_cts_width;
 		tb_y = sl->slice_segment_addr / pic_in_cts_width;
 
-		tile_block = tb_y << 12 | tb_x;
+		tile_block = AVD_OP_SL_LOC_Y(tb_y) | AVD_OP_SL_LOC_X(tb_x);
 
 		if (!(sl->flags & V4L2_HEVC_SLICE_PARAMS_FLAG_DEPENDENT_SLICE_SEGMENT))
 			last_tile_block = tile_block;
@@ -579,28 +597,29 @@ static int submit_slice_segment(struct avd_ctx *ctx, struct avd_hevc_run *run,
 		 * tile block start
 		 * CABAC window
 		 */
-		push(0x2c000000 | last_tile_block, "cm3_cmd_set_cabac_xy");
+		push(AVD_OP_SL_LOC | last_tile_block, "cm3_cmd_set_cabac_xy");
 
 		stream_slice_dqtblk(ctx, run, sl);
 	} else {
-		tile_boundary = row_bd[row] << 12 | col_bd[col];
+		tile_boundary = AVD_OP_SL_LOC_Y(row_bd[row]) |
+			        AVD_OP_SL_LOC_X(col_bd[col]);
 	}
 
 	if (coded_flags & NEW_TILE_ID) {
-		push(0x2a000000 |
-				(coded_flags & NEW_SLICE ? tile_block : tile_boundary),
+		push(AVD_OP_SL_DIM_START |
+			(coded_flags & NEW_SLICE ? tile_block : tile_boundary),
 				"cm3_cmd_set_ctb_xy");
 
 		/* tile boundary end */
 		if (pps->flags & V4L2_HEVC_PPS_FLAG_TILES_ENABLED)
-			push(((hflip ? 4 : 0) | (vflip ? 8 : 0)) << 28
-					| col << 24
-					| (row_bd[row + 1] - 1) << 12
-					| (col_bd[col + 1] - 1),
+			push(AVD_SL_DIM_END_ROW((hflip ? 4 : 0) | (vflip ? 8 : 0)) |
+					AVD_SL_DIM_END_COL(col) |
+					AVD_SL_DIM_END_Y(row_bd[row + 1] - 1) |
+					AVD_SL_DIM_END_X(col_bd[col + 1] - 1),
 					"cm3_set_ctb_xy");
 		else /* first slice, one CTB */
-			push((pic_in_cts_height - 1) << 12
-					| (pic_in_cts_width - 1),
+			push(AVD_SL_DIM_END_Y(pic_in_cts_height - 1) |
+					AVD_SL_DIM_END_X(pic_in_cts_width - 1),
 					"cm3_set_ctb_xy");
 	}
 
@@ -609,10 +628,12 @@ static int submit_slice_segment(struct avd_ctx *ctx, struct avd_hevc_run *run,
 
 	/* current tile block / boundary ?? */
 	/* Unlike entropy, motion vector window resets every time */
-	push(1 << 24 | (coded_flags & NEW_SLICE ? tile_block : tile_boundary),
+	push(AVD_SL_DIM_END_COL(1) |
+			(coded_flags & NEW_SLICE ? tile_block : tile_boundary),
 			"cm3_set_mv_xy");
 
-	push(0x2b000000 | is_last << 10, "cm3_cmd_inst_fifo_end");
+	push(AVD_OP_EXEC | AVD_OP_EXEC_FLAG_END(is_last),
+			"cm3_cmd_inst_fifo_end");
 
 	return last_tile_block;
 }
@@ -729,8 +750,10 @@ static int avd_wait_submission_queue(struct avd_ctx *ctx)
 				avd->variant->submit_queue_status_offset +
 				(ctx->vp_slot) * 4);
 
+	/* pr_info("%d/%d\n", cur, max); */
+
 	if (cur == max) {
-		dev_err_ratelimited(avd->dev, "instruction que full! %d/%d",
+		dev_err(avd->dev, "instruction que full! %d/%d",
 				    cur, max);
 		return 1;
 	}
@@ -791,12 +814,16 @@ static void stream_slices(struct avd_ctx *ctx, struct avd_hevc_run *run)
 
 	ctb_addr_rs_to_ts = kzalloc(
 		sizeof(*ctb_addr_rs_to_ts) * pic_in_ctbs_size, GFP_KERNEL);
-	if (!ctb_addr_rs_to_ts)
+	if (!ctb_addr_rs_to_ts) {
+		pr_err("alloc: ctb_addr_rs_to_ts\n");
 		goto done;
+	}
 
 	tile_ids = kzalloc(sizeof(*tile_ids) * pic_in_ctbs_size, GFP_KERNEL);
-	if (!tile_ids)
+	if (!tile_ids) {
+		pr_err("alloc: tile_ids\n");
 		goto done;
+	}
 
 	if (tiles_enabled) {
 		if (pps->flags & V4L2_HEVC_PPS_FLAG_UNIFORM_SPACING) {
@@ -830,7 +857,11 @@ static void stream_slices(struct avd_ctx *ctx, struct avd_hevc_run *run)
 		sl = &run->sl[s];
 		slice_segment_offset = 0;
 		to = tiles_enabled ? sl->num_entry_point_offsets + 1 : 1;
-
+		if (tiles_enabled && sl->num_entry_point_offsets + entry_point_idx > run->num_entry_point_offsets + 1) {
+			dev_err(ctx->dev->dev, "to few entry points! has: %d, needs > %d",
+					run->num_entry_point_offsets, sl->num_entry_point_offsets + entry_point_idx);
+			goto done;
+		}
 		for (i = 0; i < to; i++) {
 			is_last = i == to - 1 && s == run->num_slices - 1;
 			first_segment = i == 0;
@@ -924,11 +955,9 @@ static void stream_slices(struct avd_ctx *ctx, struct avd_hevc_run *run)
 	hevc_ctx->submit_num = pos;
 
 done:
-	if (ctb_addr_rs_to_ts)
-		kfree(ctb_addr_rs_to_ts);
+	kfree(ctb_addr_rs_to_ts);
 
-	if (tile_ids)
-		kfree(tile_ids);
+	kfree(tile_ids);
 }
 
 static void update_dec_buf_info(struct avd_decoded_buffer *buf,
@@ -1049,7 +1078,7 @@ static void avd_hevc_stop(struct avd_ctx *ctx)
 	kfree(hevc_ctx);
 }
 
-static void avd_hevc_run_preamble(struct avd_ctx *ctx, struct avd_hevc_run *run)
+static int avd_hevc_run_preamble(struct avd_ctx *ctx, struct avd_hevc_run *run)
 {
 	struct v4l2_ctrl *ctrl;
 	u32 dst_len, sps_len;
@@ -1072,6 +1101,9 @@ static void avd_hevc_run_preamble(struct avd_ctx *ctx, struct avd_hevc_run *run)
 	ctrl = v4l2_ctrl_find(&ctx->ctrl_hdl,
 			      V4L2_CID_STATELESS_HEVC_ENTRY_POINT_OFFSETS);
 	run->entry_point_offsets = ctrl ? ctrl->p_cur.p : NULL;
+	if (!run->entry_point_offsets)
+		return -EINVAL;
+	run->num_entry_point_offsets = ctrl ? ctrl->new_elems : 0;
 
 	avd_run_preamble(ctx, &run->base);
 
@@ -1094,6 +1126,7 @@ static void avd_hevc_run_preamble(struct avd_ctx *ctx, struct avd_hevc_run *run)
 		run->addresses.y + (dst_len - sps_len - ctx->rvra.size);
 
 	run->addresses.sps = run->addresses.y + (dst_len - sps_len);
+	return 0;
 }
 
 static int avd_hevc_run(struct avd_ctx *ctx)
@@ -1104,9 +1137,12 @@ static int avd_hevc_run(struct avd_ctx *ctx)
 	struct avd_decoded_buffer *dst;
 	int ret;
 
-	avd_hevc_run_preamble(ctx, &run);
+	ret = avd_hevc_run_preamble(ctx, &run);
+	if (ret) {
+		dev_err(avd->dev, "avd_hevc_run_preamble: failed %d", ret);
+		return ret;
+	}
 
-	avd_run_postamble(ctx, &run.base);
 	hevc_ctx = ctx->priv;
 	dst = vb2_to_avd_decoded_buf(&run.base.bufs.dst->vb2_buf);
 	update_dec_buf_info(dst, &run.sl[0]);
@@ -1116,12 +1152,16 @@ static int avd_hevc_run(struct avd_ctx *ctx)
 		dev_err(avd->dev, "no free slots: %d", ret);
 		return ret;
 	}
+	/* pr_info("VP%d: start\n", ctx->vp_slot); */
 	avd->variant->configure_stream(avd, hevc_ctx->bufs.inst.addr,
 				       ctx->fifo_idx, ctx->vp_slot);
+	/* avd_status(avd, ctx->vp_slot); */
 	set_header(ctx, &run);
 
 	schedule_delayed_work(&ctx->watchdog_work, msecs_to_jiffies(2000));
 	stream_slices(ctx, &run);
+	/* avd_status(avd, ctx->vp_slot); */
+	avd_run_postamble(ctx, &run.base);
 
 	return 0;
 }
diff --git a/drivers/media/platform/apple/avd/avd-inst.h b/drivers/media/platform/apple/avd/avd-inst.h
index 78d7e37f60bac8..8ee19f65e768e5 100644
--- a/drivers/media/platform/apple/avd/avd-inst.h
+++ b/drivers/media/platform/apple/avd/avd-inst.h
@@ -12,6 +12,137 @@
 #define INST_DMA2 0 /* (0x4000000 | INST_DMA1) */
 #define INST_DMA3 0 /* (0x07 << 16 | 0x07) */
 
+#define AVD_OP_EXEC			FIELD_PREP(GENMASK(31, 24), 0x2b)
+#define AVD_OP_EXEC_FIFO_MASK(v)	FIELD_PREP(GENMASK(3, 0), v)
+#define AVD_OP_EXEC_FIFO_IDX(v)		FIELD_PREP(GENMASK(7, 4), v)
+#define AVD_OP_EXEC_FLAG_START_REV3(v)	FIELD_PREP(BIT(8), !!(v))
+#define AVD_OP_EXEC_FLAG_START_REV4(v)	FIELD_PREP(BIT(9), !!(v))
+#define AVD_OP_EXEC_FLAG_END(v)		FIELD_PREP(BIT(10), !!(v))
+#define AVD_OP_EXEC_REV3_VP9_MASK	FIELD_PREP(GENMASK(23, 12), 0xfff)
+
+#define AVD_OP_HDR			FIELD_PREP(GENMASK(31, 20), 0x2db)
+#define AVD_OP_HDR_CONST		FIELD_PREP(GENMASK(10, 0), 0x2e0)
+#define AVD_OP_HDR_FLAG0		FIELD_PREP(BIT(12), 1)
+#define AVD_OP_HDR_FLAG_INTRA(v)	FIELD_PREP(BIT(13), !!(v))
+#define AVD_OP_HDR_FLAG_PIPE_STATE(v)	FIELD_PREP(BIT(19), !!(v))
+
+#define AVD_OP_WEIGHTS_HDR		FIELD_PREP(GENMASK(31, 20), 0x2dd)
+#define AVD_OP_WEIGHTS_HDR_CHROMA(v)	FIELD_PREP(GENMASK(2, 0), v)
+#define AVD_OP_WEIGHTS_HDR_LUMA(v)	FIELD_PREP(GENMASK(5, 3), v)
+#define AVD_OP_WEIGHTS_HDR_FLAG0(v)	FIELD_PREP(BIT(6), !!(v))
+#define AVD_OP_WEIGHTS_HDR_FLAG1(v)	FIELD_PREP(BIT(7), !!(v))
+
+#define AVD_OP_WEIGHTS			FIELD_PREP(GENMASK(31, 20), 0x2de)
+#define AVD_OP_WEIGHTS_WEIGHT(v)	FIELD_PREP(GENMASK(8, 0), v)
+#define AVD_OP_WEIGHTS_INDEX(v)		FIELD_PREP(GENMASK(12, 9), v)
+#define AVD_OP_WEIGHTS_LIST_IDX(v)	FIELD_PREP(BIT(13), v)
+/* 1 = luma, 2,3 = chroma[{0,1}] */
+#define AVD_OP_WEIGHTS_IDENT(v)		FIELD_PREP(GENMASK(16, 14), v)
+
+#define AVD_OP_OFFSETS			FIELD_PREP(GENMASK(31, 20), 0x2df)
+#define AVD_OP_OFFSETS_OFFSET(v)	FIELD_PREP(GENMASK(15, 0), v)
+
+#define AVD_OP_CODED_DATA		FIELD_PREP(GENMASK(31, 20), 0x2d8)
+#define AVD_OP_CODED_DATA_ADDR(v)	FIELD_PREP(GENMASK(12, 0), v)
+#define AVD_OP_CODED_DATA_FLAG0(v)	FIELD_PREP(BIT(13), !!(v))
+#define AVD_OP_CODED_DATA_FLAG1(v)	FIELD_PREP(BIT(14), !!(v))
+#define AVD_OP_CODED_DATA_BIT_OFF(v)	FIELD_PREP(GENMASK(18, 15), v)
+
+/* not super happy with xy */
+#define AVD_OP_SL_LOC			FIELD_PREP(GENMASK(31, 24), 0x2c)
+#define AVD_OP_SL_LOC_X(v)		FIELD_PREP(GENMASK(11, 0), v)
+#define AVD_OP_SL_LOC_Y(v)		FIELD_PREP(GENMASK(23, 12), v)
+
+#define AVD_OP_SL_DIM_START		FIELD_PREP(GENMASK(31, 24), 0x2a)
+#define AVD_OP_SL_DIM_START_X(v)	FIELD_PREP(GENMASK(11, 0), v)
+#define AVD_OP_SL_DIM_START_Y(v)	FIELD_PREP(GENMASK(23, 12), v)
+
+/* end is not an op */
+#define AVD_SL_DIM_END_X(v)		FIELD_PREP(GENMASK(11, 0), v)
+#define AVD_SL_DIM_END_Y(v)		FIELD_PREP(GENMASK(23, 12), v)
+#define AVD_SL_DIM_END_COL(v)		FIELD_PREP(GENMASK(27, 24), v)
+#define AVD_SL_DIM_END_ROW(v)		FIELD_PREP(GENMASK(31, 28), v)
+
+#define AVD_OP_SL_REF			FIELD_PREP(GENMASK(31, 24), 0x2d)
+#define AVD_OP_SL_REF_MAX_MERGE(v)	FIELD_PREP(GENMASK(3, 1), v)
+
+#define AVD_OP_SL_REF_FLAG0(v)		FIELD_PREP(BIT(4), !!(v))
+#define AVD_OP_SL_REF_FLAG_CABAC(v)	FIELD_PREP(BIT(5), !!(v))
+#define AVD_OP_SL_REF_FLAG1(v)		FIELD_PREP(BIT(6), !!(v))
+#define AVD_OP_SL_REF_NUM_L0(v)		FIELD_PREP(GENMASK(15, 11), v)
+#define AVD_OP_SL_REF_NUM_L1(v)		FIELD_PREP(GENMASK(10, 7), v)
+#define AVD_OP_SL_REF_FLAG2(v)		FIELD_PREP(BIT(15), !!(v))
+#define AVD_OP_SL_REF_SLICE_P(v)	FIELD_PREP(BIT(16), !!(v))
+#define AVD_OP_SL_REF_SLICE_I(v)	FIELD_PREP(BIT(17), !!(v))
+/* not really kinda more like has_ref_and_ref_is_valid_ref */
+#define AVD_OP_SL_REF_SLICE_B(v)	FIELD_PREP(BIT(18), !!(v))
+
+#define AVD_OP_QP			FIELD_PREP(GENMASK(31, 20), 0x2d9)
+#define AVD_OP_QP_CR_OFF(v)		FIELD_PREP(GENMASK(4, 0), v)
+#define AVD_OP_QP_CB_OFF(v)		FIELD_PREP(GENMASK(9, 5), v)
+#define AVD_OP_QP_VAL(v)		FIELD_PREP(GENMASK(17, 10), v)
+
+#define AVD_OP_DBLK			FIELD_PREP(GENMASK(31, 20), 0x2da)
+#define AVD_OP_DBLK_FLAG_SAO_CHROMA(v)	FIELD_PREP(BIT(6), !!(v))
+#define AVD_OP_DBLK_FLAG_SAO_LUMA(v)	FIELD_PREP(BIT(7), !!(v))
+#define AVD_OP_DBLK_OFF0(v)		FIELD_PREP(GENMASK(11, 8), v)
+#define AVD_OP_DBLK_OFF1(v)		FIELD_PREP(GENMASK(16, 12), v)
+#define AVD_OP_DBLK_FLAG_EN(v)		FIELD_PREP(BIT(16), !!(v))
+#define AVD_OP_DBLK_FLAG_FULL_EN(v)	FIELD_PREP(BIT(17), !!(v))
+#define AVD_OP_DBLK_FLAG_TILES_EN(v)	FIELD_PREP(BIT(18), !!(v))
+#define AVD_OP_DBLK_FLAG_PCM_EN(v)	FIELD_PREP(BIT(19), !!(v))
+
+#define AVD_OP_REF			FIELD_PREP(GENMASK(31, 20), 0x2dc)
+/* same order as they where submitted */
+#define AVD_OP_REF_DBP_IDX(v)		FIELD_PREP(GENMASK(3, 0), v)
+#define AVD_OP_REF_LOOP_IDX(v)		FIELD_PREP(GENMASK(7, 4), v)
+#define AVD_OP_REF_LIST_IDX(v)		FIELD_PREP(GENMASK(11, 8), v)
+
+/*
+ * now comes all common / shared
+ */
+#define AVD_HDR_CODEC_MODE(v)		FIELD_PREP(GENMASK(28, 24), v)
+#define AVD_HDR_WIDTH(v)		FIELD_PREP(GENMASK(15, 0), v)
+#define AVD_HDR_HEIGHT(v)		FIELD_PREP(GENMASK(31, 16), v)
+
+#define AVD_HDR_FEAT_H264		FIELD_PREP(GENMASK(3, 0), 10)
+#define AVD_HDR_FEAT_PIPE_STATE_EN(v)	FIELD_PREP(GENMASK(7, 4), (v) ? 3 : 0)
+#define AVD_HDR_FEAT_H26X		FIELD_PREP(BIT(20), 1)
+#define AVD_HDR_FEAT_COMMON		FIELD_PREP(BIT(21), 1)
+#define AVD_HDR_FEAT_VP9		FIELD_PREP(BIT(17), 1)
+
+
+#define AVD_HDR_COMMON_FLAG0(v)		FIELD_PREP(BIT(0), !!(v))
+/*
+ * this is hevc specifik
+ * TODO
+ * some of this is likeley wrong, it wont work without the overlap...
+ */
+#define AVD_HDR_COMMON_TR_INTRA(v)	FIELD_PREP(GENMASK(4, 1), v)
+#define AVD_HDR_COMMON_TR_INTER(v)	FIELD_PREP(GENMASK(7, 4), v)
+/* luma {transform,coding} block size */
+#define AVD_HDR_COMMON_LUMA_TBS(v)	FIELD_PREP(GENMASK(8, 7), v)
+#define AVD_HDR_COMMON_MIN_LUMA_TBS(v)	FIELD_PREP(GENMASK(10, 9), v)
+#define AVD_HDR_COMMON_LUMA_CBS(v)	FIELD_PREP(GENMASK(12, 11), v)
+#define AVD_HDR_COMMON_MIN_LUMA_CBS(v)	FIELD_PREP(GENMASK(14, 13), v)
+#define AVD_HDR_COMMON_BIT_DEPTH_L(v)	FIELD_PREP(GENMASK(18, 15), v)
+#define AVD_HDR_COMMON_BIT_DEPTH_C(v)	FIELD_PREP(GENMASK(23, 19), v)
+/* TODO: did not bother checking max chroma idc */
+#define AVD_HDR_COMMON_CHROMA_FORMAT(v)	FIELD_PREP(GENMASK(26, 24), v)
+
+#define AVD_HDR_H26X_QP_OFFSET_CR(v)	FIELD_PREP(GENMASK(4, 0), v)
+#define AVD_HDR_H26X_QP_OFFSET_CB(v)	FIELD_PREP(GENMASK(9, 5), v)
+
+#define AVD_SCALING_I0(v)		FIELD_PREP(GENMASK(7, 0), v)
+#define AVD_SCALING_I1(v)		FIELD_PREP(GENMASK(15, 8), v)
+#define AVD_SCALING_I2(v)		FIELD_PREP(GENMASK(23, 16), v)
+#define AVD_SCALING_I3(v)		FIELD_PREP(GENMASK(31, 24), v)
+
+#define AVD_REF_NUM(v)			FIELD_PREP(GENMASK(31, 28), v)
+#define AVD_REF_FLAG_CONST		FIELD_PREP(BIT(24), 1)
+#define AVD_REF_FLAG_LONG(v)		FIELD_PREP(BIT(17), !!(v))
+#define AVD_REF_DELTA_POC(v)		FIELD_PREP(GENMASK(16, 0), v)
+
 static inline u32 fifo_size(void)
 {
 	return 0x100000 * 12;
diff --git a/drivers/media/platform/apple/avd/avd-vp9.c b/drivers/media/platform/apple/avd/avd-vp9.c
index 77dd49de6cf140..a6230c59c772ea 100644
--- a/drivers/media/platform/apple/avd/avd-vp9.c
+++ b/drivers/media/platform/apple/avd/avd-vp9.c
@@ -24,6 +24,23 @@
 #include "avd.h"
 #include "avd-inst.h"
 
+#define VP9_Q_IDX(v)		FIELD_PREP(GENMASK(31, 15), v)
+#define VP9_Q_DC_Y(v)		FIELD_PREP(GENMASK(14, 10), v)
+#define VP9_Q_DC_UV(v)		FIELD_PREP(GENMASK(9, 5), v)
+#define VP9_Q_AC_UV(v)		FIELD_PREP(GENMASK(4, 0), v)
+
+#define VP9_LF_SHARPNESS(v)	FIELD_PREP(GENMASK(31, 28), v)
+#define VP9_LF_REF0(v)		FIELD_PREP(GENMASK(27, 21), v)
+#define VP9_LF_REF1(v)		FIELD_PREP(GENMASK(20, 14), v)
+#define VP9_LF_REF2(v)		FIELD_PREP(GENMASK(13, 7), v)
+#define VP9_LF_REF3(v)		FIELD_PREP(GENMASK(6, 0), v)
+
+#define VP9_LF_LV(v)		FIELD_PREP(GENMASK(31, 14), v)
+#define VP9_LF_MODE0(v)		FIELD_PREP(GENMASK(13, 7), v)
+#define VP9_LF_MODE1(v)		FIELD_PREP(GENMASK(6, 0), v)
+
+#define VP9_FEAT_LVL_ALT_Q(v)		FIELD_PREP(GENMASK(21, 12), v)
+
 struct avd_vp9_seg_probs {
 	u8 tree_probs[7];
 	u8 pred_probs[3];
@@ -177,8 +194,9 @@ static void set_refs(struct avd_ctx *ctx, struct avd_vp9_run *run)
 			+ (ref_buf[i]->base.vb.planes[0].length - ref_buf[i]->rvra.size);
 
 		/* TODO */
-		push(0x1000000, "hdr_9c_ref_100");
-		push((ref_buf[i]->vp9.height - 1) << 16 | (ref_buf[i]->vp9.width - 1),
+		push(AVD_REF_FLAG_CONST, "hdr_9c_ref_100");
+		push(AVD_HDR_HEIGHT(ref_buf[i]->vp9.height - 1) |
+				AVD_HDR_WIDTH(ref_buf[i]->vp9.width - 1),
 				"hdr_70_ref_height_width");
 		push(0x40004000, "hdr_7c_ref_align");
 
@@ -186,6 +204,7 @@ static void set_refs(struct avd_ctx *ctx, struct avd_vp9_run *run)
 	}
 }
 
+/* TODO */
 static u32 make_flags1(struct avd_ctx *ctx, struct avd_vp9_run *run)
 {
 	struct avd_vp9_ctx *vp9_ctx = ctx->priv;
@@ -244,7 +263,7 @@ static u32 seg_features(struct avd_ctx *ctx, struct avd_vp9_run *run,
 	if (v4l2_vp9_seg_feat_enabled(seg->feature_enabled, feature_id, segid))
 		enabled |= 1;
 
-	return (feature_val & 0x3ff) << 12 | enabled;
+	return VP9_FEAT_LVL_ALT_Q(feature_val) | enabled;
 }
 
 static void set_header(struct avd_ctx *ctx, struct avd_vp9_run *run)
@@ -259,33 +278,36 @@ static void set_header(struct avd_ctx *ctx, struct avd_vp9_run *run)
 			(V4L2_VP9_FRAME_FLAG_KEY_FRAME |
 			 V4L2_VP9_FRAME_FLAG_INTRA_ONLY));
 
-	push(0x2b000000
-			| (avd->variant->revision == 3 ? 0xfff100 : 0x200)
-			| (ctx->fifo_idx << 4),
+	push(AVD_OP_EXEC
+			| AVD_OP_EXEC_FLAG_START_REV3(avd->variant->revision == 3)
+			| AVD_OP_EXEC_FLAG_START_REV4(avd->variant->revision == 4)
+			| (avd->variant->revision == 3 ? AVD_OP_EXEC_REV3_VP9_MASK : 0)
+			| AVD_OP_EXEC_FIFO_IDX(ctx->fifo_idx),
 			"inst_fifo_start");
 
-	push(0x2db00000
-			| 0x1000
-			| ((intra_only) ? 0x2000 : 0)
-			| 0x2e0
-			| (avd->variant->quirks & AVD_QUIRK_NO_PIPE_STATE ? 0 : 0x80000)
+	push(AVD_OP_HDR
+			| AVD_OP_HDR_FLAG0
+			| AVD_OP_HDR_FLAG_INTRA(intra_only)
+			| AVD_OP_HDR_CONST
+			| AVD_OP_HDR_FLAG_PIPE_STATE(!(avd->variant->quirks & AVD_QUIRK_NO_PIPE_STATE))
 			, "hdr_34_start_hdr");
 
-	push(AVD_CODEC_VP9 << 24, "hdr_38_mode");
+	push(AVD_HDR_CODEC_MODE(AVD_CODEC_VP9), "hdr_38_mode");
 
-	push(((frame->frame_height_minus_1) << 16) | (frame->frame_width_minus_1),
+	push(AVD_HDR_HEIGHT(frame->frame_height_minus_1) |
+			AVD_HDR_WIDTH(frame->frame_width_minus_1),
 			"hdr_28_height_width_shift3");
 	push(0, "cm3_dma_config_0");
-	push(((frame->frame_height_minus_1) << 16) | (frame->frame_width_minus_1),
+	push(AVD_HDR_HEIGHT(frame->frame_height_minus_1) |
+			AVD_HDR_WIDTH(frame->frame_width_minus_1),
 			"hdr_38_height_width_shift3");
 
-	push(0x1000000 /* (sps->chroma_format_idc & 3) << 25 ? */
-			| frame->profile << 19
-			| (frame->bit_depth - 8) << 15
-			| 0x1000 /* colorspace? */
-			| 0x800
-			| (min(prob_updates->tx_mode, 3) << 7)
-			| boolify(prob_updates->tx_mode & V4L2_VP9_TX_MODE_SELECT)
+	push(AVD_HDR_COMMON_CHROMA_FORMAT(1)
+			| AVD_HDR_COMMON_BIT_DEPTH_C(frame->profile)
+			| AVD_HDR_COMMON_BIT_DEPTH_L(frame->bit_depth - 8)
+			| AVD_HDR_COMMON_LUMA_CBS(3)
+			| AVD_HDR_COMMON_LUMA_TBS(min(prob_updates->tx_mode, 3))
+			| AVD_HDR_COMMON_FLAG0(prob_updates->tx_mode & V4L2_VP9_TX_MODE_SELECT)
 			,"hdr_2c_txfm_mode");
 
 	push(make_flags1(ctx, run), "hdr_40_flags1_pt1");
@@ -294,7 +316,7 @@ static void set_header(struct avd_ctx *ctx, struct avd_vp9_run *run)
 		push(seg_features(ctx, run, i), "seg");
 
 	/* some kind of feature enable? h26{2,5} has 3 instread */
-	push(0x20000, "unk_const");
+	push(AVD_HDR_FEAT_VP9, "unk_const");
 	push(INST_DMA2, "cm3_dma_config_2");
 	push(INST_DMA1, "cm3_dma_config_3");
 
@@ -311,24 +333,24 @@ static void set_header(struct avd_ctx *ctx, struct avd_vp9_run *run)
 	pusha(vp9_ctx->bufs.above_info.addr, "hdr_110_pps2_tile_addr_lsb8", 3);
 	pusha(vp9_ctx->bufs.above_info.addr, "hdr_110_pps2_tile_addr_lsb8", 4);
 
-	push(frame->quant.base_q_idx << 15
-			| (frame->quant.delta_q_y_dc & 0x1f) << 10
-			| (frame->quant.delta_q_uv_dc & 0x1f) << 5
-			| (frame->quant.delta_q_uv_ac & 0x1f)
+	push(VP9_Q_IDX(frame->quant.base_q_idx)
+			| VP9_Q_DC_Y(frame->quant.delta_q_y_dc)
+			| VP9_Q_DC_UV(frame->quant.delta_q_uv_dc)
+			| VP9_Q_AC_UV(frame->quant.delta_q_uv_ac)
 			, "hdr_4c_base_q_idx");
 	/* filter related flags? */
-	push(frame->lf.sharpness << 28
+	push(VP9_LF_SHARPNESS(frame->lf.sharpness)
 			| (frame->lf.flags & V4L2_VP9_LOOP_FILTER_FLAG_DELTA_ENABLED ?
-				(frame->lf.ref_deltas[0] & 0x7f) << 21
-				| (frame->lf.ref_deltas[1] & 0x7f) << 14
-				| (frame->lf.ref_deltas[2] & 0x7f) << 7
-				| (frame->lf.ref_deltas[3] & 0x7f)
+				VP9_LF_REF0(frame->lf.ref_deltas[0])
+				| VP9_LF_REF1(frame->lf.ref_deltas[1])
+				| VP9_LF_REF2(frame->lf.ref_deltas[2])
+				| VP9_LF_REF3(frame->lf.ref_deltas[3])
 				: 0), "hdr_44_flags1_pt2");
 
-	push(frame->lf.level << 14
+	push(VP9_LF_LV(frame->lf.level)
 			| (frame->lf.flags & V4L2_VP9_LOOP_FILTER_FLAG_DELTA_ENABLED ?
-			(frame->lf.mode_deltas[0] & 0x7f) << 7
-			| (frame->lf.mode_deltas[1] & 0x7f)
+			VP9_LF_MODE0(frame->lf.mode_deltas[0])
+			| VP9_LF_MODE1(frame->lf.mode_deltas[1])
 			: 0),
 			"hdr_48_loop_filter_level");
 
@@ -367,7 +389,8 @@ static void set_header(struct avd_ctx *ctx, struct avd_vp9_run *run)
 	pusha(run->addresses.uv, "hdr_16c_uv_addr_lsb8", 0);
 	push(bytesperline, "hdr_174_width_align");
 	push(0, "");
-	push(((frame->frame_height_minus_1) << 16) | (frame->frame_width_minus_1),
+	push(AVD_HDR_HEIGHT(frame->frame_height_minus_1) |
+			AVD_HDR_WIDTH(frame->frame_width_minus_1),
 			"cm3_height_width");
 
 	if (!(intra_only))
@@ -408,22 +431,24 @@ static void set_tiles(struct avd_ctx *ctx, struct avd_vp9_run *run)
 				offset += 4;
 				size -= 4;
 			}
-			push(0x2d800000 | (u32)(run->addresses.sl >> 32),
+			push(AVD_OP_CODED_DATA |
+					AVD_OP_CODED_DATA_ADDR(run->addresses.sl >> 32),
 					"cm3_cmd_set_slice_data");
 			push((u32)((run->addresses.sl + offset) & 0xffffffff),
 					"til_ab4_tile_addr_low");
 			push(tile_size, "til_ab8_tile_size");
-			push(0x2a000000
-					| ((row * sb_64_rows) / num_tile_rows) << 12
-					| (col * sb_64_cols) / num_tile_cols, "i");
+			push(AVD_OP_SL_DIM_START
+					| AVD_OP_SL_DIM_START_Y((row * sb_64_rows) / num_tile_rows)
+					| AVD_OP_SL_DIM_START_X((col * sb_64_cols) / num_tile_cols), "i");
 
-			push(col << 24
-					| (((row + 1) * sb_64_rows) / num_tile_rows - 1) << 12
-					| (((col + 1) * sb_64_cols) / num_tile_cols - 1)
+			push(AVD_SL_DIM_END_COL(col)
+					| AVD_SL_DIM_END_Y(((row + 1) * sb_64_rows) / num_tile_rows - 1)
+					| AVD_SL_DIM_END_X(((col + 1) * sb_64_cols) / num_tile_cols - 1)
 					, "til_ac0_tile_dims");
-			push(0x2b000000
-					| (is_last ? 0x400 : 0)
-					| (avd->variant->revision == 3 ? 0xfff000 : 0)
+			push(AVD_OP_EXEC
+					| AVD_OP_EXEC_FLAG_END(is_last)
+					| (avd->variant->revision == 3 ?
+						AVD_OP_EXEC_REV3_VP9_MASK : 0)
 					, "cm3_cmd_inst_fifo_end");
 			offset += tile_size;
 			size -= tile_size;
@@ -630,6 +655,20 @@ static int avd_vp9_alloc_bufs(struct avd_ctx *ctx)
 	if (ret)
 		return ret;
 
+#ifdef AVD_DEBUG
+#define DBG_BUF(name, buf) pr_info("   %10s %12llx %8lx", \
+		name, vp9_ctx->bufs.buf.addr, vp9_ctx->bufs.buf.size)
+	pr_info("buffers:");
+	DBG_BUF("state", state);
+	DBG_BUF("above_info", above_info);
+	DBG_BUF("seg", seg);
+	DBG_BUF("tiles[0]", tiles[0]);
+	DBG_BUF("tiles[1]", tiles[1]);
+	DBG_BUF("tiles[2]", tiles[2]);
+	DBG_BUF("color[0]", color[0]);
+	DBG_BUF("color[1]", color[1]);
+	pr_info("\n");
+#endif
 	return 0;
 }
 
@@ -746,6 +785,7 @@ static void avd_vp9_done(struct avd_ctx *ctx,
 {
 	struct avd_vp9_ctx *vp9_ctx = ctx->priv;
 	unsigned int fctx_idx;
+	avd_status(ctx->dev, 8);
 
 	/* v4l2-specific stuff */
 	if (result == VB2_BUF_STATE_ERROR)
