diff --git a/drivers/media/platform/apple/avd/avd-hevc.c b/drivers/media/platform/apple/avd/avd-hevc.c
index e7dbe65f3bd6b2..478db2d9eca076 100644
--- a/drivers/media/platform/apple/avd/avd-hevc.c
+++ b/drivers/media/platform/apple/avd/avd-hevc.c
@@ -311,7 +311,7 @@ static void set_header(struct avd_ctx *ctx, struct avd_hevc_run *run)
 
 	push(0, "cm3_mark_end_section");
 
-	if (avd->variant->revision != 3)
+	if (!(avd->variant->quirks & AVD_QUIRK_LSR))
 		push(0, "cm3_mark_end_section");
 
 	bytesperline = ctx->decoded_fmt.fmt.pix_mp.plane_fmt[0].bytesperline;
